// @webex/calling/dist/Events/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.WEBSOCKET_SCOPE = exports.WEBSOCKET_KEYS = exports.SessionType = exports.SUPPLEMENTARY_SERVICES = exports.MOBIUS_MIDCALL_STATE = exports.MOBIUS_EVENT_KEYS = exports.MEDIA_CONNECTION_EVENT_KEYS = exports.LINE_EVENT_KEYS = exports.Disposition = exports.COMMON_EVENT_KEYS = exports.CALL_EVENT_KEYS = exports.CALLING_CLIENT_EVENT_KEYS = void 0;
var _types = require("../CallingClient/line/types");
/* eslint-disable @typescript-eslint/no-unused-vars */
// TODO: remove once we define the payloads
/** External Eventing Start */
var COMMON_EVENT_KEYS = exports.COMMON_EVENT_KEYS = /*#__PURE__*/function (COMMON_EVENT_KEYS) {
  COMMON_EVENT_KEYS["CB_VOICEMESSAGE_CONTENT_GET"] = "call_back_voicemail_content_get";
  COMMON_EVENT_KEYS["CALL_HISTORY_USER_SESSION_INFO"] = "callHistory:user_recent_sessions";
  COMMON_EVENT_KEYS["CALL_HISTORY_USER_VIEWED_SESSIONS"] = "callHistory:user_viewed_sessions";
  COMMON_EVENT_KEYS["CALL_HISTORY_USER_SESSIONS_DELETED"] = "callHistory:user_sessions_deleted";
  return COMMON_EVENT_KEYS;
}({});
var LINE_EVENT_KEYS = exports.LINE_EVENT_KEYS = /*#__PURE__*/function (LINE_EVENT_KEYS) {
  LINE_EVENT_KEYS["INCOMING_CALL"] = "incoming_call";
  return LINE_EVENT_KEYS;
}({});
var CALLING_CLIENT_EVENT_KEYS = exports.CALLING_CLIENT_EVENT_KEYS = /*#__PURE__*/function (CALLING_CLIENT_EVENT_KEYS) {
  CALLING_CLIENT_EVENT_KEYS["ERROR"] = "callingClient:error";
  CALLING_CLIENT_EVENT_KEYS["OUTGOING_CALL"] = "callingClient:outgoing_call";
  CALLING_CLIENT_EVENT_KEYS["USER_SESSION_INFO"] = "callingClient:user_recent_sessions";
  CALLING_CLIENT_EVENT_KEYS["ALL_CALLS_CLEARED"] = "callingClient:all_calls_cleared";
  return CALLING_CLIENT_EVENT_KEYS;
}({});
var CALL_EVENT_KEYS = exports.CALL_EVENT_KEYS = /*#__PURE__*/function (CALL_EVENT_KEYS) {
  CALL_EVENT_KEYS["ALERTING"] = "alerting";
  CALL_EVENT_KEYS["CALL_ERROR"] = "call_error";
  CALL_EVENT_KEYS["CALLER_ID"] = "caller_id";
  CALL_EVENT_KEYS["CONNECT"] = "connect";
  CALL_EVENT_KEYS["DISCONNECT"] = "disconnect";
  CALL_EVENT_KEYS["ESTABLISHED"] = "established";
  CALL_EVENT_KEYS["HELD"] = "held";
  CALL_EVENT_KEYS["HOLD_ERROR"] = "hold_error";
  CALL_EVENT_KEYS["PROGRESS"] = "progress";
  CALL_EVENT_KEYS["REMOTE_MEDIA"] = "remote_media";
  CALL_EVENT_KEYS["RESUME_ERROR"] = "resume_error";
  CALL_EVENT_KEYS["RESUMED"] = "resumed";
  CALL_EVENT_KEYS["TRANSFER_ERROR"] = "transfer_error";
  return CALL_EVENT_KEYS;
}({});
var SUPPLEMENTARY_SERVICES = exports.SUPPLEMENTARY_SERVICES = /*#__PURE__*/function (SUPPLEMENTARY_SERVICES) {
  SUPPLEMENTARY_SERVICES["HOLD"] = "hold";
  SUPPLEMENTARY_SERVICES["RESUME"] = "resume";
  SUPPLEMENTARY_SERVICES["DIVERT"] = "divert";
  SUPPLEMENTARY_SERVICES["TRANSFER"] = "transfer";
  SUPPLEMENTARY_SERVICES["PARK"] = "park";
  return SUPPLEMENTARY_SERVICES;
}({});
var MOBIUS_MIDCALL_STATE = exports.MOBIUS_MIDCALL_STATE = /*#__PURE__*/function (MOBIUS_MIDCALL_STATE) {
  MOBIUS_MIDCALL_STATE["HELD"] = "HELD";
  MOBIUS_MIDCALL_STATE["CONNECTED"] = "CONNECTED";
  return MOBIUS_MIDCALL_STATE;
}({});
var Disposition = exports.Disposition = /*#__PURE__*/function (Disposition) {
  Disposition["ANSWERED"] = "Answered";
  Disposition["CANCELED"] = "Canceled";
  Disposition["INITIATED"] = "Initiated";
  Disposition["MISSED"] = "MISSED";
  return Disposition;
}({});
var SessionType = exports.SessionType = /*#__PURE__*/function (SessionType) {
  SessionType["SPARK"] = "SPARK";
  SessionType["WEBEX_CALLING"] = "WEBEXCALLING";
  return SessionType;
}({});
var MOBIUS_EVENT_KEYS = exports.MOBIUS_EVENT_KEYS = /*#__PURE__*/function (MOBIUS_EVENT_KEYS) {
  MOBIUS_EVENT_KEYS["SERVER_EVENT_INCLUSIVE"] = "event:mobius";
  MOBIUS_EVENT_KEYS["CALL_SESSION_EVENT_INCLUSIVE"] = "event:janus.user_recent_sessions";
  MOBIUS_EVENT_KEYS["CALL_SESSION_EVENT_LEGACY"] = "event:janus.user_sessions";
  MOBIUS_EVENT_KEYS["CALL_SESSION_EVENT_VIEWED"] = "event:janus.user_viewed_sessions";
  MOBIUS_EVENT_KEYS["CALL_SESSION_EVENT_DELETED"] = "event:janus.user_sessions_deleted";
  return MOBIUS_EVENT_KEYS;
}({});
var MEDIA_CONNECTION_EVENT_KEYS = exports.MEDIA_CONNECTION_EVENT_KEYS = /*#__PURE__*/function (MEDIA_CONNECTION_EVENT_KEYS) {
  MEDIA_CONNECTION_EVENT_KEYS["ROAP_MESSAGE_TO_SEND"] = "roap:messageToSend";
  MEDIA_CONNECTION_EVENT_KEYS["MEDIA_TYPE_AUDIO"] = "audio";
  return MEDIA_CONNECTION_EVENT_KEYS;
}({});
/* External Eventing End */
/** Internal Eventing Start */
var CALL_STATE = /*#__PURE__*/function (CALL_STATE) {
  CALL_STATE["HELD"] = "held";
  CALL_STATE["REMOTE_HELD"] = "remoteheld";
  CALL_STATE["CONNECTED"] = "connected";
  return CALL_STATE;
}(CALL_STATE || {});
var WEBSOCKET_SCOPE = exports.WEBSOCKET_SCOPE = 'mobius';
var WEBSOCKET_KEYS = exports.WEBSOCKET_KEYS = /*#__PURE__*/function (WEBSOCKET_KEYS) {
  WEBSOCKET_KEYS["CALL_PROGRESS"] = "callprogress";
  WEBSOCKET_KEYS["CALL_CONNECTED"] = "callconnected";
  WEBSOCKET_KEYS["CALL_DISCONNECTED"] = "callconnected";
  WEBSOCKET_KEYS["CALL_INFO"] = "callinfo";
  WEBSOCKET_KEYS["CALL"] = "call";
  WEBSOCKET_KEYS["ROAP"] = "ROAP";
  return WEBSOCKET_KEYS;
}({});
/** Internal Eventing End */
/** State Machine Events */
/* ROAP Events */
// TODO: export RoapMessage type from the media SDK to use here
// eslint-disable-next-line @typescript-eslint/no-explicit-any
