// @webex/calling/dist/Logger/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.LOG_PREFIX = exports.LOGGING_LEVEL = exports.LOGGER = void 0;
/* eslint-disable @typescript-eslint/no-empty-interface */
var LOG_PREFIX = exports.LOG_PREFIX = /*#__PURE__*/function (LOG_PREFIX) {
  LOG_PREFIX["MAIN"] = "CALLING_SDK";
  LOG_PREFIX["FILE"] = "file";
  LOG_PREFIX["METHOD"] = "method";
  LOG_PREFIX["EVENT"] = "event";
  LOG_PREFIX["MESSAGE"] = "message";
  LOG_PREFIX["ERROR"] = "error";
  return LOG_PREFIX;
}({});
var LOGGING_LEVEL = exports.LOGGING_LEVEL = /*#__PURE__*/function (LOGGING_LEVEL) {
  LOGGING_LEVEL[LOGGING_LEVEL["error"] = 1] = "error";
  LOGGING_LEVEL[LOGGING_LEVEL["warn"] = 2] = "warn";
  LOGGING_LEVEL[LOGGING_LEVEL["log"] = 3] = "log";
  LOGGING_LEVEL[LOGGING_LEVEL["info"] = 4] = "info";
  LOGGING_LEVEL[LOGGING_LEVEL["trace"] = 5] = "trace";
  return LOGGING_LEVEL;
}({});
var LOGGER = exports.LOGGER = /*#__PURE__*/function (LOGGER) {
  LOGGER["ERROR"] = "error";
  LOGGER["WARN"] = "warn";
  LOGGER["INFO"] = "info";
  LOGGER["LOG"] = "log";
  LOGGER["TRACE"] = "trace";
  return LOGGER;
}({});
