// @webex/calling/dist/CallingClient/callingClientFixtures.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.uri = exports.registrationPayload = exports.regionPayload = exports.regionBody = exports.primaryUrl = exports.myIP = exports.mockUSServiceHosts = exports.mockIntServiceHosts = exports.mockEUServiceHosts = exports.mockEUIntServiceHosts = exports.mockCatalogUSInt = exports.mockCatalogUS = exports.mockCatalogEUInt = exports.mockCatalogEU = exports.ipPayload = exports.discoveryPayload = void 0;
var _testUtil = require("../common/testUtil");
var _constants = require("./constants");
var _registerFixtures = require("./registration/registerFixtures");
var webex = (0, _testUtil.getTestUtilsWebex)();
var mockIPReturnBody = {
  ipv4: '1.1.1.1',
  ipv6: '2.2.2.2'
};
var ipPayload = exports.ipPayload = {
  statusCode: 200,
  body: mockIPReturnBody
};
var regionBody = exports.regionBody = {
  attribution: 'This product includes GeoLite2 data created by MaxMind, available from http://www.maxmind.com',
  clientAddress: '72.163.220.6',
  clientRegion: 'AP-SOUTHEAST',
  countryCode: 'IN',
  disclaimer: 'This service is intended for use by Webex Team only. Unauthorized use is prohibited.',
  regionCode: 'AP-SOUTHEAST',
  timezone: 'Asia/Kolkata'
};
var regionPayload = exports.regionPayload = {
  statusCode: 200,
  body: regionBody
};
var discoveryBody = (0, _testUtil.getMobiusDiscoveryResponse)();
var primaryUrl = exports.primaryUrl = "".concat(discoveryBody.primary.uris[0], "/calling/web/");
var discoveryPayload = exports.discoveryPayload = {
  statusCode: 200,
  body: discoveryBody
};
var registrationPayload = exports.registrationPayload = {
  statusCode: 200,
  body: _registerFixtures.mockPostResponse
};
var uri = exports.uri = "".concat(webex.internal.services._serviceUrls.mobius).concat(_constants.URL_ENDPOINT);
var myIP = exports.myIP = mockIPReturnBody.ipv4;
var mockUSServiceHosts = exports.mockUSServiceHosts = [{
  host: 'mobius-us-east-1.prod.infra.webex.com',
  ttl: -1,
  priority: 5,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ca-central-1.prod.infra.webex.com',
  ttl: -1,
  priority: 10,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-eu-central-1.prod.infra.webex.com',
  ttl: -1,
  priority: 15,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ap-southeast-2.prod.infra.webex.com',
  ttl: -1,
  priority: 20,
  id: 'urn:TEAM:us-east-2_a:mobius'
}];
var mockEUServiceHosts = exports.mockEUServiceHosts = [{
  host: 'mobius-eu-central-1.prod.infra.webex.com',
  ttl: -1,
  priority: 5,
  id: 'urn:TEAM:eu-central-1_k:mobius'
}, {
  host: 'mobius-us-east-1.prod.infra.webex.com',
  ttl: -1,
  priority: 15,
  id: 'urn:TEAM:eu-central-1_k:mobius'
}, {
  host: 'mobius-ca-central-1.prod.infra.webex.com',
  ttl: -1,
  priority: 10,
  id: 'urn:TEAM:eu-central-1_k:mobius'
}, {
  host: 'mobius-ap-southeast-2.prod.infra.webex.com',
  ttl: -1,
  priority: 20,
  id: 'urn:TEAM:eu-central-1_k:mobius'
}];
var mockEUIntServiceHosts = exports.mockEUIntServiceHosts = [{
  host: 'mobius-eu-central-1.int.infra.webex.com',
  ttl: -1,
  priority: 15,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-us-east-1.int.infra.webex.com',
  ttl: -1,
  priority: 5,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ca-central-1.int.infra.webex.com',
  ttl: -1,
  priority: 10,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ap-southeast-2.int.infra.webex.com',
  ttl: -1,
  priority: 20,
  id: 'urn:TEAM:us-east-2_a:mobius'
}];
var mockIntServiceHosts = exports.mockIntServiceHosts = [{
  host: 'mobius-us-east-1.int.infra.webex.com',
  ttl: -1,
  priority: 5,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ca-central-1.int.infra.webex.com',
  ttl: -1,
  priority: 10,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-eu-central-1.int.infra.webex.com',
  ttl: -1,
  priority: 15,
  id: 'urn:TEAM:us-east-2_a:mobius'
}, {
  host: 'mobius-ap-southeast-2.int.infra.webex.com',
  ttl: -1,
  priority: 20,
  id: 'urn:TEAM:us-east-2_a:mobius'
}];
var mockCatalogEU = exports.mockCatalogEU = {
  'mobius-eu-central-1.prod.infra.webex.com': mockEUServiceHosts
};
var mockCatalogEUInt = exports.mockCatalogEUInt = {
  'mobius-eu-central-1.int.infra.webex.com': mockEUIntServiceHosts
};
var mockCatalogUS = exports.mockCatalogUS = {
  'mobius-us-east-1.prod.infra.webex.com': mockUSServiceHosts,
  'mobius-eu-central-1.prod.infra.webex.com': mockEUServiceHosts
};
var mockCatalogUSInt = exports.mockCatalogUSInt = {
  'mobius-us-east-1.int.infra.webex.com': mockIntServiceHosts,
  'mobius-eu-central-1.int.infra.webex.com': mockEUIntServiceHosts
};
