// @webex/calling/dist/CallingClient/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.WEBEX_WEB_CLIENT = exports.WEBEX = exports.VERSION = exports.VALID_PHONE_REGEX = exports.UTILS_FILE = exports.USER_AGENT_VERSION = exports.URL_ENDPOINT = exports.TYPE = exports.TRANSPORT = exports.TRANSFER_ENDPOINT = exports.TOTAL_SAMPLES_DURATION = exports.TOTAL_ROUND_TRIP_TIME = exports.TIMESTAMP = exports.TARGET_BIT_RATE = exports.SUPPLEMENTARY_SERVICES_TIMEOUT = exports.SPARK_USER_AGENT = exports.SELECTED_CANDIDATE_PAIR_ID = exports.SEC_TO_MSEC_MFACTOR = exports.RTP_TX_STAT = exports.RTP_RX_STAT = exports.RTC_ICE_CANDIDATE_PAIR = exports.RTC_ICE_CANDIDATE = exports.RTC_CODEC = exports.ROUND_TRIP_TIME_MEASUREMENTS = exports.RETRY_TIMER_UPPER_LIMIT = exports.RESUME_ENDPOINT = exports.REPO_NAME = exports.REMOTE_INBOUND_RTP = exports.REG_TRY_BACKUP_TIMER_VAL_IN_SEC = exports.REG_TRY_BACKUP_TIMER_VAL_FOR_CC_IN_SEC = exports.REG_RANDOM_T_FACTOR_UPPER_LIMIT = exports.REG_FAILBACK_429_MAX_RETRIES = exports.REG_429_RETRY_UTIL = exports.REGISTRATION_FILE = exports.REGISTER_UTIL = exports.REGISTER_RETRY_TIMEOUT = exports.RECONNECT_UTIL = exports.PACKETS_SENT = exports.PACKETS_RECEIVED = exports.PACKETS_LOST = exports.PACKETS_DISCARDED = exports.OUTBOUND_RTP = exports.OUTBOUND_CODEC_MATCH = exports.NOISE_REDUCTION_EFFECT = exports.NETWORK_TYPE = exports.NETWORK_FLAP_TIMEOUT = exports.NETWORK_CHANGE_DETECTION_UTIL = exports.MOBIUS_US_PROD = exports.MOBIUS_US_INT = exports.MOBIUS_EU_PROD = exports.MOBIUS_EU_INT = exports.MINUTES_TO_SEC_MFACTOR = exports.MIME_TYPE = exports.METRIC_FILE = exports.METHODS = exports.MEDIA_SOURCE = exports.MEDIA_ID = exports.MEDIA_ENDPOINT_RESOURCE = exports.LOCAL_CANDIDATE_ID = exports.LINE_FILE = exports.KEEPALIVE_UTIL = exports.JITTER_BUFFER_EMITTED_COUNT = exports.JITTER_BUFFER_DELAY = exports.IP_ENDPOINT = exports.INITIAL_SEQ_NUMBER = exports.INBOUND_RTP = exports.INBOUND_CODEC_MATCH = exports.ICE_CANDIDATES_TIMEOUT = exports.HOLD_ENDPOINT = exports.GET_MOBIUS_SERVERS_UTIL = exports.FAILOVER_UTIL = exports.FAILBACK_UTIL = exports.DUMMY_MOBIUS_URL = exports.DUMMY_METRICS = exports.DISCOVERY_URL = exports.DEVICES_ENDPOINT_RESOURCE = exports.DEFAULT_SESSION_TIMER = exports.DEFAULT_REHOMING_INTERVAL_MIN = exports.DEFAULT_REHOMING_INTERVAL_MAX = exports.DEFAULT_LOCAL_CALL_ID = exports.DEFAULT_KEEPALIVE_INTERVAL = exports.CRLF = exports.CODEC_ID = exports.CISCO_DEVICE_URL = exports.CALL_TRANSFER_SERVICE = exports.CALL_STATUS_RESOURCE = exports.CALL_MANAGER_FILE = exports.CALL_HOLD_SERVICE = exports.CALL_FILE = exports.CALL_ENDPOINT_RESOURCE = exports.CALLS_ENDPOINT_RESOURCE = exports.CALLS_CLEARED_HANDLER_UTIL = exports.CALLING_USER_AGENT = exports.CALLING_CLIENT_FILE = exports.CALLER_ID_FILE = exports.BYTES_SENT = exports.BYTES_RECEIVED = exports.BASE_REG_TIMER_MFACTOR = exports.BASE_REG_RETRY_TIMER_VAL_IN_SEC = exports.API_V1 = void 0;
exports.WEB_AGENT = void 0;
var VERSION = exports.VERSION = 'unknown';
var USER_AGENT_VERSION = exports.USER_AGENT_VERSION = 'beta';
var REPO_NAME = exports.REPO_NAME = 'webex-calling';
var CALLING_USER_AGENT = exports.CALLING_USER_AGENT = "".concat(REPO_NAME, "/").concat(USER_AGENT_VERSION);
var CALL_ENDPOINT_RESOURCE = exports.CALL_ENDPOINT_RESOURCE = 'call';
var CALL_STATUS_RESOURCE = exports.CALL_STATUS_RESOURCE = 'status';
var CALLS_ENDPOINT_RESOURCE = exports.CALLS_ENDPOINT_RESOURCE = 'calls';
var CISCO_DEVICE_URL = exports.CISCO_DEVICE_URL = 'cisco-device-url';
var CRLF = exports.CRLF = '\r\n';
var DEFAULT_KEEPALIVE_INTERVAL = exports.DEFAULT_KEEPALIVE_INTERVAL = 30;
var DEFAULT_LOCAL_CALL_ID = exports.DEFAULT_LOCAL_CALL_ID = 'DefaultLocalId';
var DEFAULT_REHOMING_INTERVAL_MAX = exports.DEFAULT_REHOMING_INTERVAL_MAX = 120;
var DEFAULT_REHOMING_INTERVAL_MIN = exports.DEFAULT_REHOMING_INTERVAL_MIN = 60;
var DEFAULT_SESSION_TIMER = exports.DEFAULT_SESSION_TIMER = 1000 * 60 * 10;
var DEVICES_ENDPOINT_RESOURCE = exports.DEVICES_ENDPOINT_RESOURCE = 'devices';
var DISCOVERY_URL = exports.DISCOVERY_URL = 'https://ds.ciscospark.com/v1/region';
var DUMMY_METRICS = exports.DUMMY_METRICS = {
  'rtp-rxstat': {
    Dur: 0,
    Pkt: 0,
    Oct: 0,
    LatePkt: 0,
    LostPkt: 0,
    AvgJit: 0,
    VQMetrics: {
      VoRxCodec: 'unknown',
      VoPktSizeMs: 0,
      maxJitter: 0,
      VoOneWayDelayMs: 0,
      networkType: 'unknown',
      hwType: 'unknown'
    }
  },
  'rtp-txstat': {
    Dur: 0,
    Pkt: 0,
    Oct: 0,
    VQMetrics: {
      VoTxCodec: 'unknown',
      rtpBitRate: 0
    }
  }
};
var DUMMY_MOBIUS_URL = exports.DUMMY_MOBIUS_URL = 'https://mobius.aintgen-a-1.int.infra.webex.com/api/v1';
var IP_ENDPOINT = exports.IP_ENDPOINT = 'myip';
var INITIAL_SEQ_NUMBER = exports.INITIAL_SEQ_NUMBER = 1;
var MEDIA_ENDPOINT_RESOURCE = exports.MEDIA_ENDPOINT_RESOURCE = 'media';
var NETWORK_FLAP_TIMEOUT = exports.NETWORK_FLAP_TIMEOUT = 2000;
var CALL_HOLD_SERVICE = exports.CALL_HOLD_SERVICE = 'callhold';
var CALL_TRANSFER_SERVICE = exports.CALL_TRANSFER_SERVICE = 'calltransfer';
var HOLD_ENDPOINT = exports.HOLD_ENDPOINT = 'hold';
var TRANSFER_ENDPOINT = exports.TRANSFER_ENDPOINT = 'commit';
var RESUME_ENDPOINT = exports.RESUME_ENDPOINT = 'resume';
var SPARK_USER_AGENT = exports.SPARK_USER_AGENT = 'spark-user-agent';
var REGISTER_RETRY_TIMEOUT = exports.REGISTER_RETRY_TIMEOUT = 10000;
var SUPPLEMENTARY_SERVICES_TIMEOUT = exports.SUPPLEMENTARY_SERVICES_TIMEOUT = 10000;
var API_V1 = exports.API_V1 = '/api/v1';
var URL_ENDPOINT = exports.URL_ENDPOINT = '/calling/web/';
var VALID_PHONE_REGEX = exports.VALID_PHONE_REGEX = /[\d\s()*#+.-]+/;
var WEB_AGENT = exports.WEB_AGENT = '(web)';
var WEBEX = exports.WEBEX = 'webex';
var WEBEX_WEB_CLIENT = exports.WEBEX_WEB_CLIENT = 'webex-web-client';
var CALLER_ID_FILE = exports.CALLER_ID_FILE = 'CallerId';
var UTILS_FILE = exports.UTILS_FILE = 'utils';
var CALLING_CLIENT_FILE = exports.CALLING_CLIENT_FILE = 'CallingClient';
var LINE_FILE = exports.LINE_FILE = 'line';
var CALL_FILE = exports.CALL_FILE = 'call';
var CALL_MANAGER_FILE = exports.CALL_MANAGER_FILE = 'callManager';
var METRIC_FILE = exports.METRIC_FILE = 'metric';
var REGISTRATION_FILE = exports.REGISTRATION_FILE = 'register';
var CODEC_ID = exports.CODEC_ID = 'codecId';
var MEDIA_ID = exports.MEDIA_ID = 'id';
var RTC_ICE_CANDIDATE_PAIR = exports.RTC_ICE_CANDIDATE_PAIR = 'RTCIceCandidatePair_';
var LOCAL_CANDIDATE_ID = exports.LOCAL_CANDIDATE_ID = 'localCandidateId';
var RTC_ICE_CANDIDATE = exports.RTC_ICE_CANDIDATE = 'RTCIceCandidate_';
var NETWORK_TYPE = exports.NETWORK_TYPE = 'networkType';
var RTC_CODEC = exports.RTC_CODEC = 'RTCCodec_';
var INBOUND_CODEC_MATCH = exports.INBOUND_CODEC_MATCH = 'CIT01_';
var OUTBOUND_CODEC_MATCH = exports.OUTBOUND_CODEC_MATCH = 'COT01_';
var MIME_TYPE = exports.MIME_TYPE = 'mimeType';
var REMOTE_INBOUND_RTP = exports.REMOTE_INBOUND_RTP = 'remote-inbound-rtp';
var TOTAL_ROUND_TRIP_TIME = exports.TOTAL_ROUND_TRIP_TIME = 'totalRoundTripTime';
var ROUND_TRIP_TIME_MEASUREMENTS = exports.ROUND_TRIP_TIME_MEASUREMENTS = 'roundTripTimeMeasurements';
var INBOUND_RTP = exports.INBOUND_RTP = 'inbound-rtp';
var OUTBOUND_RTP = exports.OUTBOUND_RTP = 'outbound-rtp';
var PACKETS_RECEIVED = exports.PACKETS_RECEIVED = 'packetsReceived';
var PACKETS_SENT = exports.PACKETS_SENT = 'packetsSent';
var PACKETS_LOST = exports.PACKETS_LOST = 'packetsLost';
var PACKETS_DISCARDED = exports.PACKETS_DISCARDED = 'packetsDiscarded';
var JITTER_BUFFER_DELAY = exports.JITTER_BUFFER_DELAY = 'jitterBufferDelay';
var JITTER_BUFFER_EMITTED_COUNT = exports.JITTER_BUFFER_EMITTED_COUNT = 'jitterBufferEmittedCount';
var TIMESTAMP = exports.TIMESTAMP = 'timestamp';
var TYPE = exports.TYPE = 'type';
var TRANSPORT = exports.TRANSPORT = 'transport';
var TARGET_BIT_RATE = exports.TARGET_BIT_RATE = 'targetBitrate';
var MEDIA_SOURCE = exports.MEDIA_SOURCE = 'media-source';
var BYTES_RECEIVED = exports.BYTES_RECEIVED = 'bytesReceived';
var BYTES_SENT = exports.BYTES_SENT = 'bytesSent';
var SELECTED_CANDIDATE_PAIR_ID = exports.SELECTED_CANDIDATE_PAIR_ID = 'selectedCandidatePairId';
var TOTAL_SAMPLES_DURATION = exports.TOTAL_SAMPLES_DURATION = 'totalSamplesDuration';
var RTP_RX_STAT = exports.RTP_RX_STAT = 'rtp-rxstat';
var RTP_TX_STAT = exports.RTP_TX_STAT = 'rtp-txstat';
var BASE_REG_TIMER_MFACTOR = exports.BASE_REG_TIMER_MFACTOR = 2;
var BASE_REG_RETRY_TIMER_VAL_IN_SEC = exports.BASE_REG_RETRY_TIMER_VAL_IN_SEC = 30;
var SEC_TO_MSEC_MFACTOR = exports.SEC_TO_MSEC_MFACTOR = 1000;
var MINUTES_TO_SEC_MFACTOR = exports.MINUTES_TO_SEC_MFACTOR = 60;
var REG_RANDOM_T_FACTOR_UPPER_LIMIT = exports.REG_RANDOM_T_FACTOR_UPPER_LIMIT = 10000;
var REG_TRY_BACKUP_TIMER_VAL_IN_SEC = exports.REG_TRY_BACKUP_TIMER_VAL_IN_SEC = 1200;
var REG_TRY_BACKUP_TIMER_VAL_FOR_CC_IN_SEC = exports.REG_TRY_BACKUP_TIMER_VAL_FOR_CC_IN_SEC = 114;
var REG_FAILBACK_429_MAX_RETRIES = exports.REG_FAILBACK_429_MAX_RETRIES = 5;
var RETRY_TIMER_UPPER_LIMIT = exports.RETRY_TIMER_UPPER_LIMIT = 60;
var REGISTER_UTIL = exports.REGISTER_UTIL = 'attemptRegistrationWithServers';
var GET_MOBIUS_SERVERS_UTIL = exports.GET_MOBIUS_SERVERS_UTIL = 'getMobiusServers';
var KEEPALIVE_UTIL = exports.KEEPALIVE_UTIL = 'startKeepaliveTimer';
var FAILBACK_UTIL = exports.FAILBACK_UTIL = 'executeFailback';
var REG_429_RETRY_UTIL = exports.REG_429_RETRY_UTIL = 'handle429Retry';
var FAILOVER_UTIL = exports.FAILOVER_UTIL = 'startFailoverTimer';
var NETWORK_CHANGE_DETECTION_UTIL = exports.NETWORK_CHANGE_DETECTION_UTIL = 'detectNetworkChange';
var CALLS_CLEARED_HANDLER_UTIL = exports.CALLS_CLEARED_HANDLER_UTIL = 'callsClearedHandler';
var RECONNECT_UTIL = exports.RECONNECT_UTIL = 'reconnectOnFailure';
var NOISE_REDUCTION_EFFECT = exports.NOISE_REDUCTION_EFFECT = 'noise-reduction-effect';
var MOBIUS_US_PROD = exports.MOBIUS_US_PROD = 'mobius-us-east-1.prod.infra.webex.com';
var MOBIUS_EU_PROD = exports.MOBIUS_EU_PROD = 'mobius-eu-central-1.prod.infra.webex.com';
var MOBIUS_US_INT = exports.MOBIUS_US_INT = 'mobius-us-east-1.int.infra.webex.com';
var MOBIUS_EU_INT = exports.MOBIUS_EU_INT = 'mobius-eu-central-1.int.infra.webex.com';
var ICE_CANDIDATES_TIMEOUT = exports.ICE_CANDIDATES_TIMEOUT = 3000;
// Define constants for method names
var METHODS = exports.METHODS = {
  CONSTRUCTOR: 'constructor',
  CREATE_CALL: 'createCall',
  HANDLE_INCOMING_CALL_SETUP: 'handleIncomingCallSetup',
  HANDLE_OUTGOING_CALL_SETUP: 'handleOutgoingCallSetup',
  HANDLE_CALL_HOLD: 'handleCallHold',
  HANDLE_CALL_RESUME: 'handleCallResume',
  HANDLE_INCOMING_CALL_PROGRESS: 'handleIncomingCallProgress',
  HANDLE_INCOMING_ROAP_OFFER_REQUEST: 'handleIncomingRoapOfferRequest',
  HANDLE_OUTGOING_CALL_ALERTING: 'handleOutgoingCallAlerting',
  HANDLE_INCOMING_CALL_CONNECT: 'handleIncomingCallConnect',
  HANDLE_OUTGOING_CALL_CONNECT: 'handleOutgoingCallConnect',
  HANDLE_INCOMING_CALL_DISCONNECT: 'handleIncomingCallDisconnect',
  HANDLE_OUTGOING_CALL_DISCONNECT: 'handleOutgoingCallDisconnect',
  HANDLE_CALL_ESTABLISHED: 'handleCallEstablished',
  HANDLE_UNKNOWN_STATE: 'handleUnknownState',
  HANDLE_TIMEOUT: 'handleTimeout',
  GET_EMITTER_CALLBACK: 'getEmitterCallback',
  HANDLE_ROAP_ESTABLISHED: 'handleRoapEstablished',
  HANDLE_ROAP_ERROR: 'handleRoapError',
  HANDLE_OUTGOING_ROAP_OFFER: 'handleOutgoingRoapOffer',
  HANDLE_OUTGOING_ROAP_ANSWER: 'handleOutgoingRoapAnswer',
  HANDLE_INCOMING_ROAP_OFFER: 'handleIncomingRoapOffer',
  HANDLE_INCOMING_ROAP_ANSWER: 'handleIncomingRoapAnswer',
  FORCE_SEND_STATS_REPORT: 'forceSendStatsReport',
  UPDATE_ACTIVE_MOBIUS: 'updateActiveMobius',
  DEQUEUE_WS_EVENTS: 'dequeueWsEvents',
  GET_CALL: 'getCall',
  GET_ACTIVE_CALLS: 'getActiveCalls',
  UPDATE_LINE: 'updateLine',
  GET_LINE_ID: 'getLineId',
  INIT_MEDIA_CONNECTION: 'initMediaConnection',
  GET_DIRECTION: 'getDirection',
  GET_CALL_ID: 'getCallId',
  GET_CORRELATION_ID: 'getCorrelationId',
  SEND_CALL_STATE_MACHINE_EVT: 'sendCallStateMachineEvt',
  SEND_MEDIA_STATE_MACHINE_EVT: 'sendMediaStateMachineEvt',
  SET_CALL_ID: 'setCallId',
  SET_DISCONNECT_REASON: 'setDisconnectReason',
  GET_DISCONNECT_REASON: 'getDisconnectReason',
  ANSWER: 'answer',
  DIAL: 'dial',
  POST: 'post',
  PATCH: 'patch',
  POST_SS_REQUEST: 'postSSRequest',
  POST_STATUS: 'postStatus',
  COMPLETE_TRANSFER: 'completeTransfer',
  GET_CALL_STATS: 'getCallStats',
  POST_MEDIA: 'postMedia',
  MEDIA_ROAP_EVENTS_LISTENER: 'mediaRoapEventsListener',
  MEDIA_TRACK_LISTENER: 'mediaTrackListener',
  ON_EFFECT_ENABLED: 'onEffectEnabled',
  ON_EFFECT_DISABLED: 'onEffectDisabled',
  UPDATE_TRACK: 'updateTrack',
  REGISTER_EFFECT_LISTENER: 'registerEffectListener',
  UNREGISTER_LISTENERS: 'unregisterListeners',
  REGISTER_LISTENERS: 'registerListeners',
  DELETE: 'delete',
  SUBMIT_CALL_ERROR_METRIC: 'submitCallErrorMetric',
  HANDLE_MID_CALL_EVENT: 'handleMidCallEvent',
  GET_CALLER_INFO: 'getCallerInfo',
  END: 'end',
  DO_HOLD_RESUME: 'doHoldResume',
  START_CALLER_ID_RESOLUTION: 'startCallerIdResolution',
  SEND_DIGIT: 'sendDigit',
  MUTE: 'mute',
  IS_MUTED: 'isMuted',
  IS_CONNECTED: 'isConnected',
  IS_HELD: 'isHeld',
  UPDATE_MEDIA: 'updateMedia',
  SET_BROADWORKS_CORRELATION_INFO: 'setBroadworksCorrelationInfo',
  GET_BROADWORKS_CORRELATION_INFO: 'getBroadworksCorrelationInfo',
  GET_CALL_RTP_STATS: 'getCallRtpStats',
  REGISTER: 'register',
  DEREGISTER: 'deregister',
  NORMALIZE_LINE: 'normalizeLine',
  LINE_EMITTER: 'lineEmitter',
  GET_LOGGING_LEVEL: 'getLoggingLevel',
  GET_ACTIVE_MOBIUS_URL: 'getActiveMobiusUrl',
  GET_STATUS: 'getStatus',
  GET_DEVICE_ID: 'getDeviceId',
  MAKE_CALL: 'makeCall',
  INCOMING_CALL_LISTENER: 'incomingCallListener',
  SET_MOBIUS_SERVERS: 'setMobiusServers',
  HANDLE_CONNECTION_RESTORATION: 'handleConnectionRestoration',
  RECONNECT_ON_FAILURE: 'reconnectOnFailure',
  DETECT_NETWORK_CHANGE: 'detectNetworkChange',
  GET_CLIENT_REGION_INFO: 'getClientRegionInfo',
  GET_MOBIUS_SERVERS: 'getMobiusServers',
  REGISTER_CALLS_CLEARED_LISTENER: 'registerCallsClearedListener',
  CALLS_CLEARED_HANDLER: 'callsClearedHandler',
  REGISTER_SESSIONS_LISTENER: 'registerSessionsListener',
  CREATE_LINE: 'createLine',
  GET_LINES: 'getLines',
  UPLOAD_LOGS: 'uploadLogs',
  GET_SDK_CONNECTOR: 'getSDKConnector',
  GET_CONNECTED_CALL: 'getConnectedCall'
};
