// @webex/calling/dist/CallingClient/calling/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferType = exports.RoapScenario = exports.MobiusEventType = exports.MobiusCallState = exports.MidCallEventType = exports.MediaState = exports.MUTE_TYPE = exports.DisconnectCode = exports.DisconnectCause = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
var MobiusEventType = exports.MobiusEventType = /*#__PURE__*/function (MobiusEventType) {
  MobiusEventType["CALL_SETUP"] = "mobius.call";
  MobiusEventType["CALL_PROGRESS"] = "mobius.callprogress";
  MobiusEventType["CALL_CONNECTED"] = "mobius.callconnected";
  MobiusEventType["CALL_MEDIA"] = "mobius.media";
  MobiusEventType["CALL_DISCONNECTED"] = "mobius.calldisconnected";
  return MobiusEventType;
}({});
var MediaState = exports.MediaState = /*#__PURE__*/function (MediaState) {
  MediaState["OFFER"] = "OFFER";
  MediaState["ANSWER"] = "ANSWER";
  MediaState["OFFER_REQUEST"] = "OFFER_REQUEST";
  MediaState["OK"] = "OK";
  MediaState["ERROR"] = "ERROR";
  return MediaState;
}({});
var DisconnectCode = exports.DisconnectCode = /*#__PURE__*/function (DisconnectCode) {
  DisconnectCode[DisconnectCode["BUSY"] = 115] = "BUSY";
  DisconnectCode[DisconnectCode["NORMAL"] = 0] = "NORMAL";
  DisconnectCode[DisconnectCode["MEDIA_INACTIVITY"] = 131] = "MEDIA_INACTIVITY";
  return DisconnectCode;
}({});
var DisconnectCause = exports.DisconnectCause = /*#__PURE__*/function (DisconnectCause) {
  DisconnectCause["BUSY"] = "User Busy.";
  DisconnectCause["NORMAL"] = "Normal Disconnect.";
  DisconnectCause["MEDIA_INACTIVITY"] = "Media Inactivity.";
  return DisconnectCause;
}({});
/* Work in Progress */
var MidCallEventType = exports.MidCallEventType = /*#__PURE__*/function (MidCallEventType) {
  MidCallEventType["CALL_INFO"] = "callInfo";
  MidCallEventType["CALL_STATE"] = "callState";
  return MidCallEventType;
}({});
var RoapScenario = exports.RoapScenario = /*#__PURE__*/function (RoapScenario) {
  RoapScenario["ANSWER"] = "ANSWER";
  RoapScenario["OK"] = "OK";
  RoapScenario["OFFER"] = "OFFER";
  RoapScenario["ERROR"] = "ERROR";
  RoapScenario["OFFER_RESPONSE"] = "OFFER_RESPONSE";
  return RoapScenario;
}({});
var MobiusCallState = exports.MobiusCallState = /*#__PURE__*/function (MobiusCallState) {
  MobiusCallState["PROCEEDING"] = "sig_proceeding";
  MobiusCallState["PROGRESS"] = "sig_progress";
  MobiusCallState["ALERTING"] = "sig_alerting";
  MobiusCallState["CONNECTED"] = "sig_connected";
  return MobiusCallState;
}({});
var TransferType = exports.TransferType = /*#__PURE__*/function (TransferType) {
  TransferType["BLIND"] = "BLIND";
  TransferType["CONSULT"] = "CONSULT";
  return TransferType;
}({});
var MUTE_TYPE = exports.MUTE_TYPE = /*#__PURE__*/function (MUTE_TYPE) {
  MUTE_TYPE["USER"] = "user_mute";
  MUTE_TYPE["SYSTEM"] = "system_mute";
  return MUTE_TYPE;
}({});
/**
 * Represents an interface for managing call-related operations.
 */
/**
 * Represents an interface for managing calls within a call manager.
 */
