// @webex/ladon-ts/build/module/lib/pipeline.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { LocalGenerator, NullGenerator, WorkerGenerator, } from './generator';
import { BlurRenderer, PassthroughRenderer, ReplacementRenderer } from './renderer';
import { resizeImage } from './util';
export class Pipeline {
    constructor(inputCanvas, outputCanvas, config, videoElement) {
        this.inputCanvas = inputCanvas;
        this.outputCanvas = outputCanvas;
        this.config = config;
        this.videoElement = videoElement;
        this.loaded = false;
        this.frameQueue = [];
        this.maskQueue = [];
        const inputCanvasCtx = this.inputCanvas.getContext('2d', {
            alpha: false,
            willReadFrequently: true,
        });
        if (!inputCanvasCtx) {
            throw new Error('[ladon-ts] - error getting input context 2d');
        }
        this.inputCanvasCtx = inputCanvasCtx;
    }
    /**
     * Processes a single frame and returns the inference result.
     *
     * @param timestamp - The timestamp of the frame.
     * @param [performInference=true] - Flag to perform mask inference.
     * @returns The result of the frame processing.
     * @throws If the pipeline (generator or renderer) is not loaded.
     */
    onFrame(timestamp, performInference = true) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.generator || !this.renderer) {
                throw new Error('[ladon-ts] - pipeline is not loaded.');
            }
            // Get the current frame (raw image data or video element)
            const frameData = this.getInputFrameData(performInference);
            // Add the frame to the queue with its timestamp
            this.frameQueue.push({ timestamp, frameData });
            if (performInference || !this.lastMaskResult) {
                // Perform inference and add the new mask to the mask queue
                const rawImageData = frameData;
                const maskImageData = this.getMaskImageData(rawImageData);
                const maskResult = yield this.generator.getLadonOutput(maskImageData, timestamp);
                // Add the mask to the queue and use the mask timestamp to sync frames and masks.
                this.maskQueue.push({ timestamp: maskResult.timestamp, maskResult });
                this.lastMaskResult = maskResult;
                // Sync frames and masks to render together
                yield this.renderSyncedFrameAndMask();
                // Post-process the mask after rendering
                yield this.generator.postRender(maskImageData, timestamp);
            }
            else {
                // Skipping inference. Update the last mask result's timestamp to match the current frame.
                this.maskQueue.push({ timestamp, maskResult: this.lastMaskResult });
                // Sync frames and masks to render together
                yield this.renderSyncedFrameAndMask();
            }
            // Return the inference result
            return this.createInferenceResult(timestamp, performInference);
        });
    }
    /**
     * Synchronizes frames and masks to ensure they are rendered together.
     */
    renderSyncedFrameAndMask() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            while (this.frameQueue.length > 0 && this.maskQueue.length > 0) {
                const frame = this.frameQueue[0];
                const maskEntry = this.maskQueue[0];
                if (frame.timestamp === maskEntry.timestamp) {
                    // Render the frame and mask together
                    yield ((_a = this.renderer) === null || _a === void 0 ? void 0 : _a.render(frame.frameData, maskEntry.maskResult));
                    this.frameQueue.shift();
                    this.maskQueue.shift();
                }
                else if (frame.timestamp < maskEntry.timestamp) {
                    // Discard the older frame
                    this.frameQueue.shift();
                }
                else {
                    // Discard the older mask
                    this.maskQueue.shift();
                }
            }
        });
    }
    /**
     * Gets the raw image data from the input canvas context.
     *
     * @param needFreshImageData - Flag to indicate if fresh image data is needed.
     *
     * @returns The raw image data.
     */
    getInputFrameData(needFreshImageData) {
        if (needFreshImageData || !this.videoElement) {
            return this.inputCanvasCtx.getImageData(0, 0, this.config.input.width, this.config.input.height).data;
        }
        return this.videoElement;
    }
    /**
     * Resizes the image data if the model expects a fixed size input.
     *
     * @param {Uint8ClampedArray} rawImageData - The raw image data.
     * @returns {Uint8ClampedArray} - The processed mask image data.
     */
    getMaskImageData(rawImageData) {
        // Check if resizing is needed based on configuration.
        if (this.config.mask.input_size.width === this.config.input.width) {
            return rawImageData;
        }
        // Resize the image data.
        return resizeImage(rawImageData, this.config.input.width, this.config.input.height, this.config.mask.input_size.width, this.config.mask.input_size.height);
    }
    /**
     * Creates the inference result object.
     *
     * @param {number} timestamp - The timestamp of the frame.
     * @param {boolean} performedInference - Indicates if the inference was performed.
     * @returns {IFrameInferenceResult} - The result object containing inference details.
     */
    createInferenceResult(timestamp, performedInference) {
        if (!this.lastMaskResult) {
            return {
                facesAndLandmarks: { data: new Float32Array(0), dims: [] },
                gesture: new Float32Array(0),
                motion: 0,
                skipped: true,
                timestamp,
            };
        }
        return {
            facesAndLandmarks: this.lastMaskResult.facesAndLandmarks,
            gesture: this.lastMaskResult.gesture,
            motion: this.lastMaskResult.motion,
            skipped: !performedInference,
            timestamp,
        };
    }
    /**
     * Checks if the pipeline is loaded.
     *
     * @returns True if the pipeline is loaded, false otherwise.
     */
    isLoaded() {
        return this.loaded;
    }
    /**
     * Loads the pipeline components such as the generator and renderer.
     */
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            switch (this.config.render.type) {
                case 'blur':
                    this.renderer = new BlurRenderer(this.outputCanvas, this.config);
                    break;
                case 'passthrough':
                    this.renderer = new PassthroughRenderer(this.outputCanvas, this.config);
                    break;
                case 'replacement':
                    this.renderer = new ReplacementRenderer(this.outputCanvas, this.config);
                    break;
                default:
                    throw new Error('[ladon-ts] - renderer not implemented.');
            }
            yield this.renderer.load();
            if (this.config.render.type === 'passthrough') {
                // automatically disable model evaluation when the renderer is set to passthrough
                // Note: this may change when gestures are detected (model eval will be a dependency for more than rendering)
                this.generator = new NullGenerator(this.config);
            }
            else {
                switch (this.config.mask.generator) {
                    case 'local':
                        this.generator = new LocalGenerator(this.config);
                        break;
                    case 'worker':
                        this.generator = new WorkerGenerator(this.config);
                        break;
                    default:
                        throw new Error('[ladon-ts] - generator not implemented.');
                }
            }
            yield this.generator.load();
            this.loaded = true;
        });
    }
    /**
     * Resets the frame and mask queues.
     */
    resetQueues() {
        this.frameQueue = [];
        this.maskQueue = [];
    }
    reset() {
        this.resetQueues();
        this.lastMaskResult = undefined;
    }
    /**
     * Destroys the pipeline and cleans up resources.
     */
    destroy() {
        var _a, _b;
        this.reset();
        (_a = this.generator) === null || _a === void 0 ? void 0 : _a.destroy();
        (_b = this.renderer) === null || _b === void 0 ? void 0 : _b.destroy();
    }
}
