// @webex/ladon-ts/build/module/lib/shared-resource-manager.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as ort from 'onnxruntime-web';
import { toAbsoluteURI } from './util';
/**
 * Shared resource manager used to preload and share resources across different pipelines.
 * This is useful when multiple pipelines are using the same model or when the model needs
 * to be preloaded before the pipeline is created.
 */
class SharedResourceManager {
    constructor() {
        this.modelSessions = {};
    }
    /**
     * Initialize the worker using the provided configuration.
     *
     * @param {PreloadConfig} config The configuration object.
     * @returns {Promise<Worker>} A promise that resolves when the worker is initialized.
     * @example
     * await initializeWorker(config);
     */
    initializeWorker(config) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof Worker === 'undefined') {
                throw new Error('[ladon-ts] web workers are not supported in this environment');
            }
            const crossOriginScriptUrl = config.mask.worker_uri;
            const content = yield (yield fetch(crossOriginScriptUrl, { cache: 'force-cache' })).text();
            this.workerUrl = URL.createObjectURL(new Blob([content], { type: 'text/javascript' }));
            this.worker = new Worker(this.workerUrl, { type: 'module' });
            return this.worker;
        });
    }
    /**
     * Terminate the worker.
     */
    terminateWorker() {
        if (this.worker) {
            this.worker.terminate();
            this.worker = undefined;
        }
    }
    /**
     * Preload the model using the provided configuration.
     *
     * @param {PreloadConfig} config The configuration object.
     * @returns {Promise<void>} A promise that resolves when the model is preloaded.
     * @example
     * await preloadModel(config);
     */
    preloadModel(config) {
        var _a;
        var _b;
        return __awaiter(this, void 0, void 0, function* () {
            if (config.wasm_uri) {
                ort.env.wasm.wasmPaths = toAbsoluteURI(config.wasm_uri, config.base_url);
            }
            ort.env.wasm.numThreads = 0;
            // Resolve the model URL based on the configuration.
            const modelUrl = yield this.getModelUrl(config);
            // Use the model session cache to avoid creating multiple sessions for the same model.
            (_a = (_b = this.modelSessions)[modelUrl]) !== null && _a !== void 0 ? _a : (_b[modelUrl] = yield ort.InferenceSession.create(modelUrl, {
                executionProviders: ['wasm'],
            }));
            // Warmup the model with a dummy input to avoid latency when running the model for the first time.
            yield this.warmupModel(this.modelSessions[modelUrl], config.mask.input_size, config.mask.model_rank);
            return this.modelSessions[modelUrl];
        });
    }
    /**
     * Preload the worker using the provided configuration.
     *
     * @param {PreloadConfig} config The configuration object.
     * @returns {Promise<void>} A promise that resolves when the worker is preloaded.
     * @example
     * await preloadWorker(config);
     */
    preloadWorker(config) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            // If the model URL resolver is a function, We can't send the function to the worker so
            // we resolve the model URL here and send the resolved URL to the worker.
            const modelUrl = yield this.getModelUrl(config);
            const newConfig = Object.assign(Object.assign({}, config), { mask: Object.assign(Object.assign({}, config.mask), { model_url_resolver: modelUrl }) });
            // Initialize the worker.
            yield this.initializeWorker(newConfig);
            // Send the preload message to the worker.
            (_a = this.worker) === null || _a === void 0 ? void 0 : _a.postMessage({ type: 'preload', config: newConfig });
        });
    }
    /**
     * Warmup the model session with a dummy input to avoid latency when running the model for the first time.
     *
     * @param {ort.InferenceSession} session The model session.
     * @param {SizeConfig} inputSize The input size of the model.
     * @param {number} inputChannels The number of input channels (e.g., RGB is 3).
     * @param {ort.Tensor.Type} inputType The input tensor type.
     * @returns {Promise<void>} A promise that resolves when the model is warmed up.
     * @example
     * await warmupModel(session, inputSize, inputChannels, inputType);
     */
    warmupModel(session, inputSize = { width: 1280, height: 720 }, inputChannels = 3, inputType = 'float32') {
        return __awaiter(this, void 0, void 0, function* () {
            const numPixels = inputSize.height * inputSize.width;
            const dummyInputData = new Float32Array(inputChannels * numPixels).fill(0);
            const dummyInput = new ort.Tensor(inputType, dummyInputData, [
                1,
                inputChannels,
                inputSize.height,
                inputSize.width,
            ]);
            const feed = { [session.inputNames[0]]: dummyInput };
            // Use an additional dummy tensor if the model expects a previous mask.
            if (inputChannels === 3) {
                const dummyMaskData = new Float32Array(numPixels).fill(0);
                const prevMaskTensor = new ort.Tensor(inputType, dummyMaskData, [
                    1,
                    1,
                    inputSize.height,
                    inputSize.width,
                ]);
                feed[session.inputNames[1]] = prevMaskTensor;
            }
            // Run the model with the dummy input to warm it up.
            yield session.run(feed);
        });
    }
    /**
     * Get the model session for the provided model URL.
     *
     * @param {string} modelUrl The model URL.
     * @returns {ort.InferenceSession} The model session.
     * @example
     * const session = getModelSession(modelUrl);
     */
    getModelSession(modelUrl) {
        return this.modelSessions[modelUrl];
    }
    /**
     * Resolves the model URL based on the configuration.
     *
     * @param {PipelineConfig} config The configuration object.
     * @returns {Promise<string>} The resolved model URL.
     * @example
     * const modelUrl = await getModelUrl(config);
     */
    getModelUrl({ mask: { model_url_resolver, model_uri } }) {
        return __awaiter(this, void 0, void 0, function* () {
            let modelUrl;
            if (typeof model_url_resolver === 'string') {
                modelUrl = model_url_resolver;
            }
            else {
                modelUrl = yield model_url_resolver(model_uri);
            }
            return modelUrl;
        });
    }
    getWorker() {
        return this.worker;
    }
    getWorkerUrl() {
        return this.workerUrl;
    }
}
// Global instance of the shared resource manager.
export const sharedResourceManager = new SharedResourceManager();
/**
 * Preload the resources using the provided configuration. This function is used to preload the model
 * and the worker before creating the pipeline.
 *
 * @param config - The configuration object.
 */
export const preload = (config) => __awaiter(void 0, void 0, void 0, function* () {
    if (config.mask.generator === 'worker') {
        yield sharedResourceManager.preloadWorker(config);
    }
    else {
        yield sharedResourceManager.preloadModel(config);
    }
});
