// @webex/calling/dist/common/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.XSI_ACTION_ENDPOINT_ORG_URL_PARAM = exports.XSI_ACTION_ENDPOINT = exports.XML_TYPE = exports.WEBEX_CALLING_CONNECTOR_FILE = exports.WEBEX_API_PROD = exports.WEBEX_API_CONFIG_PROD_URL = exports.WEBEX_API_CONFIG_INT_URL = exports.WEBEX_API_BTS = exports.VOICEMAIL = exports.VALUES = exports.USER_SESSIONS = exports.USERS = exports.USER = exports.UNPROCESSABLE_CONTENT_CODE = exports.UCM_CONNECTOR_FILE = exports.TRANSCRIPT = exports.TOKEN = exports.TIME = exports.SUCCESS_STATUS_CODE = exports.SUCCESS_MESSAGE = exports.STATUS_CODE = exports.SETTINGS = exports.SERVICES_ENDPOINT = exports.SCIM_WEBEXIDENTITY_USER = exports.SCIM_USER_FILTER = exports.SCIM_ENTERPRISE_USER = exports.SCIM_ENDPOINT_RESOURCE = exports.RESPONSE_MESSAGE = exports.RESPONSE_DATA = exports.RESPONSE = exports.RAW_REQUEST = exports.PLACEHOLDER_KEY = exports.OBJECT = exports.NATIVE_WEBEX_TEAMS_CALLING = exports.NATIVE_SIP_CALL_TO_UCM = exports.KEY = exports.ITEMS = exports.INFER_ID_CONSTANT = exports.IDENTITY_ENDPOINT_RESOURCE = exports.FAILURE_MESSAGE = exports.ENTITLEMENT_STANDARD = exports.ENTITLEMENT_BROADWORKS_CONN = exports.ENTITLEMENT_BASIC = exports.DEVICES = exports.CONTENT = exports.BW_XSI_URL = exports.BW_XSI_ENDPOINT_VERSION_WITH_SLASH = exports.BW_XSI_ENDPOINT_VERSION = exports.BINARY = exports.BEARER = exports.BASE64 = void 0;
var BASE64 = exports.BASE64 = 'base64';
var BEARER = exports.BEARER = 'bearer';
var BINARY = exports.BINARY = 'binary';
var CONTENT = exports.CONTENT = 'content';
var DEVICES = exports.DEVICES = 'devices';
var FAILURE_MESSAGE = exports.FAILURE_MESSAGE = 'FAILURE';
var IDENTITY_ENDPOINT_RESOURCE = exports.IDENTITY_ENDPOINT_RESOURCE = 'identity';
var ITEMS = exports.ITEMS = 'items';
var KEY = exports.KEY = 'key';
var OBJECT = exports.OBJECT = 'object';
var PLACEHOLDER_KEY = exports.PLACEHOLDER_KEY = '$';
var RAW_REQUEST = exports.RAW_REQUEST = 'rawRequest';
var RESPONSE = exports.RESPONSE = 'response';
var RESPONSE_DATA = exports.RESPONSE_DATA = 'responseData';
var RESPONSE_MESSAGE = exports.RESPONSE_MESSAGE = 'responseMessage';
var SCIM_ENDPOINT_RESOURCE = exports.SCIM_ENDPOINT_RESOURCE = 'scim';
var SCIM_USER_FILTER = exports.SCIM_USER_FILTER = 'v2/Users?filter=';
var SETTINGS = exports.SETTINGS = 'settings';
var STATUS_CODE = exports.STATUS_CODE = 'statusCode';
var SUCCESS_MESSAGE = exports.SUCCESS_MESSAGE = 'SUCCESS';
var SUCCESS_STATUS_CODE = exports.SUCCESS_STATUS_CODE = 200;
var UNPROCESSABLE_CONTENT_CODE = exports.UNPROCESSABLE_CONTENT_CODE = 422;
var TIME = exports.TIME = 'time';
var TOKEN = exports.TOKEN = 'token';
var TRANSCRIPT = exports.TRANSCRIPT = 'transcript';
var USER = exports.USER = 'user';
var USERS = exports.USERS = 'users';
var USER_SESSIONS = exports.USER_SESSIONS = 'userSessions';
var VALUES = exports.VALUES = '_values';
var XML_TYPE = exports.XML_TYPE = 'text/xml';
var XSI_ACTION_ENDPOINT = exports.XSI_ACTION_ENDPOINT = 'xsiActionsEndpoint';
var XSI_ACTION_ENDPOINT_ORG_URL_PARAM = exports.XSI_ACTION_ENDPOINT_ORG_URL_PARAM = 'organizations?callingData=true';
var SERVICES_ENDPOINT = exports.SERVICES_ENDPOINT = 'services';
var INFER_ID_CONSTANT = exports.INFER_ID_CONSTANT = 'ciscospark://us';
var ENTITLEMENT_BASIC = exports.ENTITLEMENT_BASIC = 'bc-sp-basic';
var ENTITLEMENT_BROADWORKS_CONN = exports.ENTITLEMENT_BROADWORKS_CONN = 'broadworks-connector';
var ENTITLEMENT_STANDARD = exports.ENTITLEMENT_STANDARD = 'bc-sp-standard';
var NATIVE_SIP_CALL_TO_UCM = exports.NATIVE_SIP_CALL_TO_UCM = 'NATIVE_SIP_CALL_TO_UCM';
var NATIVE_WEBEX_TEAMS_CALLING = exports.NATIVE_WEBEX_TEAMS_CALLING = 'NATIVE_WEBEX_TEAMS_CALLING';
var BW_XSI_ENDPOINT_VERSION = exports.BW_XSI_ENDPOINT_VERSION = 'v2.0';
var BW_XSI_ENDPOINT_VERSION_WITH_SLASH = exports.BW_XSI_ENDPOINT_VERSION_WITH_SLASH = 'v2.0/';
var BW_XSI_URL = exports.BW_XSI_URL = 'broadworksXsiActionsUrl';
var WEBEX_CALLING_CONNECTOR_FILE = exports.WEBEX_CALLING_CONNECTOR_FILE = 'WxCallBackendConnector';
var UCM_CONNECTOR_FILE = exports.UCM_CONNECTOR_FILE = 'UcmBackendConnector';
var VOICEMAIL = exports.VOICEMAIL = 'VOICEMAIL';
var SCIM_WEBEXIDENTITY_USER = exports.SCIM_WEBEXIDENTITY_USER = 'urn:scim:schemas:extension:cisco:webexidentity:2.0:User';
var SCIM_ENTERPRISE_USER = exports.SCIM_ENTERPRISE_USER = 'urn:ietf:params:scim:schemas:extension:enterprise:2.0:User';
var WEBEX_API_PROD = exports.WEBEX_API_PROD = 'https://webexapis.com';
var WEBEX_API_BTS = exports.WEBEX_API_BTS = 'https://integration.webexapis.com';
var WEBEX_API_CONFIG_INT_URL = exports.WEBEX_API_CONFIG_INT_URL = "".concat(WEBEX_API_BTS, "/v1/uc/config");
var WEBEX_API_CONFIG_PROD_URL = exports.WEBEX_API_CONFIG_PROD_URL = "".concat(WEBEX_API_PROD, "/v1/uc/config");
