// @webex/calling/dist/common/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.ServiceIndicator = exports.SORT_BY = exports.SORT = exports.RegistrationStatus = exports.HTTP_METHODS = exports.DecodeType = exports.CallType = exports.CallDirection = exports.CALLING_BACKEND = exports.ALLOWED_SERVICES = void 0;
var _constants = require("./constants");
var ALLOWED_SERVICES = exports.ALLOWED_SERVICES = /*#__PURE__*/function (ALLOWED_SERVICES) {
  ALLOWED_SERVICES["MOBIUS"] = "mobius";
  ALLOWED_SERVICES["JANUS"] = "janus";
  return ALLOWED_SERVICES;
}({});
var HTTP_METHODS = exports.HTTP_METHODS = /*#__PURE__*/function (HTTP_METHODS) {
  HTTP_METHODS["GET"] = "GET";
  HTTP_METHODS["POST"] = "POST";
  HTTP_METHODS["PATCH"] = "PATCH";
  HTTP_METHODS["PUT"] = "PUT";
  HTTP_METHODS["DELETE"] = "DELETE";
  return HTTP_METHODS;
}({});
var RegistrationStatus = exports.RegistrationStatus = /*#__PURE__*/function (RegistrationStatus) {
  RegistrationStatus["IDLE"] = "IDLE";
  RegistrationStatus["ACTIVE"] = "active";
  RegistrationStatus["INACTIVE"] = "inactive";
  return RegistrationStatus;
}({});
var CALLING_BACKEND = exports.CALLING_BACKEND = /*#__PURE__*/function (CALLING_BACKEND) {
  CALLING_BACKEND["WXC"] = "WEBEX_CALLING";
  CALLING_BACKEND["BWRKS"] = "BROADWORKS_CALLING";
  CALLING_BACKEND["UCM"] = "UCM_CALLING";
  CALLING_BACKEND["INVALID"] = "Calling backend is currently not supported";
  return CALLING_BACKEND;
}({}); // guid;
var CallType = exports.CallType = /*#__PURE__*/function (CallType) {
  CallType["URI"] = "uri";
  CallType["TEL"] = "tel";
  return CallType;
}({});
var CallDirection = exports.CallDirection = /*#__PURE__*/function (CallDirection) {
  CallDirection["INBOUND"] = "inbound";
  CallDirection["OUTBOUND"] = "outbound";
  return CallDirection;
}({});
var SORT = exports.SORT = /*#__PURE__*/function (SORT) {
  SORT["ASC"] = "ASC";
  SORT["DESC"] = "DESC";
  SORT["DEFAULT"] = "DESC";
  return SORT;
}({});
var SORT_BY = exports.SORT_BY = /*#__PURE__*/function (SORT_BY) {
  SORT_BY["END_TIME"] = "endTime";
  SORT_BY["DEFAULT"] = "endTime";
  SORT_BY["START_TIME"] = "startTime";
  return SORT_BY;
}({});
var ServiceIndicator = exports.ServiceIndicator = /*#__PURE__*/function (ServiceIndicator) {
  ServiceIndicator["CALLING"] = "calling";
  ServiceIndicator["CONTACT_CENTER"] = "contactcenter";
  ServiceIndicator["GUEST_CALLING"] = "guestcalling";
  return ServiceIndicator;
}({});
var DecodeType = exports.DecodeType = /*#__PURE__*/function (DecodeType) {
  DecodeType["PEOPLE"] = "PEOPLE";
  DecodeType["ORGANIZATION"] = "ORGANIZATION";
  return DecodeType;
}({});
