// @webex/calling/dist/Metrics/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.VOICEMAIL_ACTION = exports.TRANSFER_ACTION = exports.REG_ACTION = exports.METRIC_TYPE = exports.METRIC_EVENT = void 0;
var METRIC_TYPE = exports.METRIC_TYPE = /*#__PURE__*/function (METRIC_TYPE) {
  METRIC_TYPE["OPERATIONAL"] = "operational";
  METRIC_TYPE["BEHAVIORAL"] = "behavioral";
  return METRIC_TYPE;
}({});
var METRIC_EVENT = exports.METRIC_EVENT = /*#__PURE__*/function (METRIC_EVENT) {
  METRIC_EVENT["BNR_ENABLED"] = "web-calling-sdk-bnr-enabled";
  METRIC_EVENT["BNR_DISABLED"] = "web-calling-sdk-bnr-disabled";
  METRIC_EVENT["CALL"] = "web-calling-sdk-callcontrol";
  METRIC_EVENT["CALL_ERROR"] = "web-calling-sdk-callcontrol-error";
  METRIC_EVENT["MEDIA"] = "web-calling-sdk-media";
  METRIC_EVENT["MEDIA_ERROR"] = "web-calling-sdk-media-error";
  METRIC_EVENT["REGISTRATION"] = "web-calling-sdk-registration";
  METRIC_EVENT["REGISTRATION_ERROR"] = "web-calling-sdk-registration-error";
  METRIC_EVENT["VOICEMAIL"] = "web-calling-sdk-voicemail";
  METRIC_EVENT["VOICEMAIL_ERROR"] = "web-calling-sdk-voicemail-error";
  return METRIC_EVENT;
}({});
var REG_ACTION = exports.REG_ACTION = /*#__PURE__*/function (REG_ACTION) {
  REG_ACTION["REGISTER"] = "register";
  REG_ACTION["DEREGISTER"] = "deregister";
  REG_ACTION["KEEPALIVE_FAILURE"] = "keepaliveFailure";
  return REG_ACTION;
}({});
var TRANSFER_ACTION = exports.TRANSFER_ACTION = /*#__PURE__*/function (TRANSFER_ACTION) {
  TRANSFER_ACTION["BLIND"] = "TRANSFER_BLIND";
  TRANSFER_ACTION["CONSULT"] = "TRANSFER_CONSULT";
  return TRANSFER_ACTION;
}({});
var VOICEMAIL_ACTION = exports.VOICEMAIL_ACTION = /*#__PURE__*/function (VOICEMAIL_ACTION) {
  VOICEMAIL_ACTION["GET_VOICEMAILS"] = "get_voicemails";
  VOICEMAIL_ACTION["GET_VOICEMAIL_CONTENT"] = "get_voicemail_content";
  VOICEMAIL_ACTION["GET_VOICEMAIL_SUMMARY"] = "get_voicemail_summary";
  VOICEMAIL_ACTION["MARK_READ"] = "mark_read";
  VOICEMAIL_ACTION["MARK_UNREAD"] = "mark_unread";
  VOICEMAIL_ACTION["DELETE"] = "delete";
  VOICEMAIL_ACTION["TRANSCRIPT"] = "transcript";
  return VOICEMAIL_ACTION;
}({});
