// @webex/ladon-ts/build/module/lib/renderer/passthrough-renderer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { compileShader, linkProgram } from '../util';
import { WebGLRenderer } from './webgl-renderer';
export class PassthroughRenderer extends WebGLRenderer {
    constructor(output_canvas, config) {
        super(output_canvas, config);
        this.output_canvas = output_canvas;
        this.config = config;
        this.fragment_shader = `#version 300 es
precision highp float;

out vec4 outColor;

uniform sampler2D       u_image;
uniform vec2            u_resolution;


void main() {
    ${this.getUVCalculation()}

    outColor = vec4(texture(u_image, st).rgb, 1.0);
}
`;
        const vertex_shader = compileShader(this.gl, this.gl.VERTEX_SHADER, this.vertex_shader_src);
        const fragment_shader = compileShader(this.gl, this.gl.FRAGMENT_SHADER, this.fragment_shader);
        this.program = linkProgram(this.gl, vertex_shader, fragment_shader);
        this.gl.useProgram(this.program);
        this.setSharedUniforms(this.program);
    }
    render(input, 
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    _) {
        return __awaiter(this, void 0, void 0, function* () {
            this.uploadVideoStream(input);
            this.gl.useProgram(this.program);
            this.gl.drawArrays(this.gl.TRIANGLES, 0, this.verts.length / 3);
            this.postDraw();
        });
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            // empty
        });
    }
    destroy() {
        // empty
    }
}
