// @webex/ladon-ts/build/module/lib/renderer/replacement-renderer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { bilinearUpsample, compileShader, linkProgram } from '../util';
import { WebGLRenderer } from './webgl-renderer';
export class ReplacementRenderer extends WebGLRenderer {
    constructor(output_canvas, config) {
        super(output_canvas, config);
        this.output_canvas = output_canvas;
        this.config = config;
        this.fragment_shader = `#version 300 es
precision highp float;

out vec4 outColor;

uniform sampler2D u_background;
uniform sampler2D u_image;
uniform sampler2D u_mask;
uniform vec2 u_resolution;


// Retrieve the mask and inverse mask
float get_mask(vec2 st) {
    float maskValue = texture(u_mask, st).r;
    // Apply a smooth gradient at the edges of the mask
    return smoothstep(0.45, 0.55, maskValue);
}

void main() {
    ${this.getUVCalculation()}

    vec3 background = texture(u_background, st).rgb;
    vec3 image = texture(u_image, st).rgb;
    float mask = get_mask(st);

    // Smooth blending between the background and the image based on the mask
    vec3 color = background * (1.0 - mask) + image * mask;
    outColor = vec4(color, 1.0);
}       
`;
        const background_canvas_ctx = this.config.render.background.getContext('2d');
        if (!background_canvas_ctx) {
            throw new Error('[ladon-ts] - could not get background canvas context 2d');
        }
        this.background_canvas_ctx = background_canvas_ctx;
        const background_data = this.background_canvas_ctx.getImageData(0, 0, this.config.render.background.width, this.config.render.background.height).data;
        if (this.config.render.kernel_size % 2 === 0) {
            throw new Error('[ladon-ts] - kernel_size must be odd');
        }
        const vertex_shader = compileShader(this.gl, this.gl.VERTEX_SHADER, this.vertex_shader_src);
        const fragment_shader = compileShader(this.gl, this.gl.FRAGMENT_SHADER, this.fragment_shader);
        this.program = linkProgram(this.gl, vertex_shader, fragment_shader);
        this.gl.useProgram(this.program);
        this.setSharedUniforms(this.program);
        const background_texture = this.gl.createTexture();
        if (!background_texture) {
            throw new Error('[ladon-ts] - error creating input texture');
        }
        this.background_texture = background_texture;
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.background_texture);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_MIN_FILTER, this.gl.LINEAR);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_S, this.gl.CLAMP_TO_EDGE);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_T, this.gl.CLAMP_TO_EDGE);
        this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.config.render.background.width, this.config.render.background.height, 0, this.gl.RGBA, this.gl.UNSIGNED_BYTE, background_data);
        this.gl.bindTexture(this.gl.TEXTURE_2D, null);
        const u_background = this.gl.getUniformLocation(this.program, 'u_background');
        if (!u_background) {
            throw new Error('[ladon-ts] - error finding "u_background" uniform');
        }
        this.gl.activeTexture(this.gl.TEXTURE2);
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.background_texture);
        this.gl.uniform1i(u_background, 2);
    }
    render(input, mask) {
        return __awaiter(this, void 0, void 0, function* () {
            this.uploadVideoStream(input);
            // The mask may not be included if it has not been generated (e.g., frame skipped).
            if (mask === null || mask === void 0 ? void 0 : mask.warm) {
                const maskData = bilinearUpsample(mask.data, this.config.mask.output_size.width, this.config.mask.output_size.height, this.config.mask.upscaled_size.width, this.config.mask.upscaled_size.height);
                this.gl.activeTexture(this.gl.TEXTURE1);
                this.gl.bindTexture(this.gl.TEXTURE_2D, this.mask_texture);
                this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.R8, this.config.mask.upscaled_size.width, this.config.mask.upscaled_size.height, 0, this.gl.RED, this.gl.UNSIGNED_BYTE, maskData);
            }
            if (!this.config.render.static) {
                const backgroundData = this.background_canvas_ctx.getImageData(0, 0, this.config.render.background.width, this.config.render.background.height).data;
                this.gl.activeTexture(this.gl.TEXTURE2);
                this.gl.bindTexture(this.gl.TEXTURE_2D, this.background_texture);
                this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.config.input.width, this.config.input.height, 0, this.gl.RGBA, this.gl.UNSIGNED_BYTE, backgroundData);
            }
            this.gl.useProgram(this.program);
            this.gl.drawArrays(this.gl.TRIANGLES, 0, this.verts.length / 3);
            this.postDraw();
        });
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            // empty
        });
    }
    destroy() {
        // empty
    }
}
