// @webex/ladon-ts/build/module/lib/renderer/webgl-renderer.js
export class WebGLRenderer {
    constructor(output_canvas, config) {
        this.output_canvas = output_canvas;
        this.config = config;
        this.verts = new Float32Array([
            -1.0, -1.0, 0.0, -1.0, 1.0, 0.0, 1.0, 1.0, 0.0,
            1.0, 1.0, 0.0, 1.0, -1.0, 0.0, -1.0, -1.0, 0.0,
        ]);
        this.vertex_shader_src = `#version 300 es

in vec4 position;

void main() {
    gl_Position = position;
}
`;
        const gl = this.output_canvas.getContext('webgl2', { antialias: true });
        if (!gl) {
            const output_2d = this.output_canvas.getContext('2d');
            if (!output_2d) {
                throw new Error('[ladon-ts] - output canvas context must be webgl2 or 2d');
            }
            this.output_canvas_2d = output_2d;
            this.intermediate_canvas = document.createElement('canvas');
            this.intermediate_canvas.height = this.config.input.height;
            this.intermediate_canvas.width = this.config.input.width;
            const gl_intermediate = this.intermediate_canvas.getContext('webgl2', {
                antialias: true,
            });
            if (!gl_intermediate) {
                throw new Error('[ladon-ts] - error getting output context webgl');
            }
            this.gl = gl_intermediate;
        }
        else {
            this.gl = gl;
        }
        // Input image texture.
        const image_texture = this.gl.createTexture();
        if (!image_texture) {
            throw new Error('[ladon-ts] - error creating input texture');
        }
        this.image_texture = image_texture;
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.image_texture);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_MIN_FILTER, this.gl.LINEAR);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_S, this.gl.CLAMP_TO_EDGE);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_T, this.gl.CLAMP_TO_EDGE);
        this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.config.input.width, this.config.input.height, 0, this.gl.RGBA, this.gl.UNSIGNED_BYTE, null);
        this.gl.bindTexture(this.gl.TEXTURE_2D, null);
        const mask_texture = this.gl.createTexture();
        if (!mask_texture) {
            throw new Error('[ladon-ts] - error creating mask texture');
        }
        // Mask texture.
        this.mask_texture = mask_texture;
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.mask_texture);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_MAG_FILTER, this.gl.LINEAR);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_MIN_FILTER, this.gl.LINEAR);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_S, this.gl.CLAMP_TO_EDGE);
        this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_T, this.gl.CLAMP_TO_EDGE);
        this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.R8, this.config.mask.upscaled_size.width, this.config.mask.upscaled_size.height, 0, this.gl.RED, this.gl.UNSIGNED_BYTE, null);
        this.gl.bindTexture(this.gl.TEXTURE_2D, null);
        const vertex_array = this.gl.createVertexArray();
        if (!vertex_array) {
            throw new Error('[ladon-ts] - error creating vertex array');
        }
        this.gl.bindVertexArray(vertex_array);
        const buffer = this.gl.createBuffer();
        if (!buffer) {
            throw new Error('[ladon-ts] - error creating buffer');
        }
        this.gl.bindBuffer(this.gl.ARRAY_BUFFER, buffer);
        this.gl.bufferData(this.gl.ARRAY_BUFFER, this.verts, this.gl.STATIC_DRAW);
        this.gl.vertexAttribPointer(0, 3, this.gl.FLOAT, false, 0, 0);
        this.gl.enableVertexAttribArray(0);
        this.gl.viewport(0, 0, this.config.input.width, this.config.input.height);
    }
    getUVCalculation() {
        if (this.config.render.horizontal_mirror) {
            return `vec2 st = -(gl_FragCoord.xy / u_resolution) + 1.0;`;
        }
        return `vec2 st = vec2((gl_FragCoord.x / u_resolution.x), -(gl_FragCoord.y / u_resolution.y) + 1.0);`;
    }
    setSharedUniforms(program) {
        const u_resolution = this.gl.getUniformLocation(program, 'u_resolution');
        if (!u_resolution) {
            throw new Error('[ladon-ts] - error finding "u_resolution" uniform');
        }
        this.gl.uniform2f(u_resolution, this.config.input.width, this.config.input.height);
        const u_image = this.gl.getUniformLocation(program, 'u_image');
        if (!u_image) {
            throw new Error('[ladon-ts] - error finding "u_image" uniform');
        }
        this.gl.activeTexture(this.gl.TEXTURE0);
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.image_texture);
        this.gl.uniform1i(u_image, 0);
        const u_mask = this.gl.getUniformLocation(program, 'u_mask');
        if (u_mask) {
            this.gl.activeTexture(this.gl.TEXTURE1);
            this.gl.bindTexture(this.gl.TEXTURE_2D, this.mask_texture);
            this.gl.uniform1i(u_mask, 1);
        }
    }
    uploadVideoStream(input) {
        this.gl.activeTexture(this.gl.TEXTURE0);
        this.gl.bindTexture(this.gl.TEXTURE_2D, this.image_texture);
        // Render video stream to texture if input is a video element.
        if (input instanceof HTMLVideoElement) {
            this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.gl.RGBA, this.gl.UNSIGNED_BYTE, input);
        }
        else {
            this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.config.input.width, this.config.input.height, 0, this.gl.RGBA, this.gl.UNSIGNED_BYTE, input);
        }
    }
    postDraw() {
        if (this.output_canvas_2d && this.intermediate_canvas) {
            this.output_canvas_2d.drawImage(this.intermediate_canvas, 0, 0, this.config.input.width, this.config.input.height);
        }
    }
}
