// @webex/calling/dist/CallSettings/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.XSI_VERSION = exports.VM_ENDPOINT = exports.USER_ENDPOINT = exports.PEOPLE_ENDPOINT = exports.ORG_ENDPOINT = exports.METHODS = exports.DND_ENDPOINT = exports.CF_ENDPOINT = exports.CALL_WAITING_ENDPOINT = exports.CALL_SETTINGS_FILE = void 0;
var CALL_SETTINGS_FILE = exports.CALL_SETTINGS_FILE = 'CallSettings';
var PEOPLE_ENDPOINT = exports.PEOPLE_ENDPOINT = 'people';
var USER_ENDPOINT = exports.USER_ENDPOINT = 'user';
var ORG_ENDPOINT = exports.ORG_ENDPOINT = 'orgId';
var DND_ENDPOINT = exports.DND_ENDPOINT = 'features/doNotDisturb';
var CF_ENDPOINT = exports.CF_ENDPOINT = 'features/callForwarding';
var VM_ENDPOINT = exports.VM_ENDPOINT = 'features/voicemail';
var CALL_WAITING_ENDPOINT = exports.CALL_WAITING_ENDPOINT = 'CallWaiting';
var XSI_VERSION = exports.XSI_VERSION = 'v2.0';

// Method name constants
var METHODS = exports.METHODS = {
  INITIALIZE_BACKEND_CONNECTOR: 'initializeBackendConnector',
  GET_CALL_WAITING_SETTING: 'getCallWaitingSetting',
  GET_DO_NOT_DISTURB_SETTING: 'getDoNotDisturbSetting',
  SET_DO_NOT_DISTURB_SETTING: 'setDoNotDisturbSetting',
  GET_CALL_FORWARD_SETTING: 'getCallForwardSetting',
  SET_CALL_FORWARD_SETTING: 'setCallForwardSetting',
  GET_VOICEMAIL_SETTING: 'getVoicemailSetting',
  SET_VOICEMAIL_SETTING: 'setVoicemailSetting',
  GET_CALL_FORWARD_ALWAYS_SETTING: 'getCallForwardAlwaysSetting',
  GET_METHOD_NOT_SUPPORTED_RESPONSE: 'getMethodNotSupportedResponse'
};
