// @webex/calling/dist/CallHistory/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.VERSION_1 = exports.USER_SESSIONS = exports.UPDATE_MISSED_CALLS_ENDPOINT = exports.UNIFIED_COMMUNICATIONS = exports.SUCCESS_MESSAGE = exports.STATUS_CODE = exports.SET_READ_STATE_SUCCESS_MESSAGE = exports.SET_DELETE_CALL_RECORDS_SUCCESS_MESSAGE = exports.SET_DELETE_CALL_RECORDS_INVALID_DATE_FORMAT_MESSAGE = exports.RESPONSE_MESSAGE = exports.PEOPLE = exports.ORG_ID = exports.NUMBER_OF_DAYS = exports.METHODS = exports.LINES = exports.LIMIT = exports.HISTORY = exports.FROM_DATE = exports.DELETE_CALL_HISTORY_RECORDS_ENDPOINT = exports.CONTENT_TYPE = exports.CONFIG = exports.CALL_HISTORY_FILE = exports.APPLICATION_JSON = void 0;
var APPLICATION_JSON = exports.APPLICATION_JSON = 'application/json';
var CALL_HISTORY_FILE = exports.CALL_HISTORY_FILE = 'CallHistory';
var CONTENT_TYPE = exports.CONTENT_TYPE = 'Content-Type';
var CONFIG = exports.CONFIG = 'config';
var DELETE_CALL_HISTORY_RECORDS_ENDPOINT = exports.DELETE_CALL_HISTORY_RECORDS_ENDPOINT = 'markAsDeleted';
var FROM_DATE = exports.FROM_DATE = '?from';
var HISTORY = exports.HISTORY = 'history';
var LIMIT = exports.LIMIT = 50;
var LINES = exports.LINES = 'lines';
var NUMBER_OF_DAYS = exports.NUMBER_OF_DAYS = 10;
var ORG_ID = exports.ORG_ID = 'orgId';
var PEOPLE = exports.PEOPLE = 'people';
var RESPONSE_MESSAGE = exports.RESPONSE_MESSAGE = 'responseMessage';
var SET_READ_STATE_SUCCESS_MESSAGE = exports.SET_READ_STATE_SUCCESS_MESSAGE = 'Missed calls are read by the user.';
var SET_DELETE_CALL_RECORDS_SUCCESS_MESSAGE = exports.SET_DELETE_CALL_RECORDS_SUCCESS_MESSAGE = 'Call history records are deleted by the user.';
var SET_DELETE_CALL_RECORDS_INVALID_DATE_FORMAT_MESSAGE = exports.SET_DELETE_CALL_RECORDS_INVALID_DATE_FORMAT_MESSAGE = 'The provided date is malformed or invalid';
var SUCCESS_MESSAGE = exports.SUCCESS_MESSAGE = 'SUCCESS';
var STATUS_CODE = exports.STATUS_CODE = 'statusCode';
var USER_SESSIONS = exports.USER_SESSIONS = 'userSessions';
var UPDATE_MISSED_CALLS_ENDPOINT = exports.UPDATE_MISSED_CALLS_ENDPOINT = 'setReadState';
var UNIFIED_COMMUNICATIONS = exports.UNIFIED_COMMUNICATIONS = 'uc';
var VERSION_1 = exports.VERSION_1 = 'v1';

// Method names
var METHODS = exports.METHODS = {
  GET_CALL_HISTORY_DATA: 'getCallHistoryData',
  UPDATE_MISSED_CALLS: 'updateMissedCalls',
  FETCH_UCM_LINES_DATA: 'fetchUCMLinesData',
  DELETE_CALL_HISTORY_RECORDS: 'deleteCallHistoryRecords'
};
