// @webex/ladon-ts/build/module/config.json.js
const e = "1.20.1", t = "0.7.7.2", i = "segmentation.js", n = { int: "https://models.int.infra.intelligence.webex.com", prod: "https://models.intelligence.webex.com" }, o = { models: "ladon/models", wasm: "ladon/wasm", workers: "ladon/workers" }, l = { weak: { kernelSize: 9, sigma: 2 }, moderate: { kernelSize: 15, sigma: 6 }, strong: { kernelSize: 25, sigma: 11 }, stronger: { kernelSize: 37, sigma: 17 }, strongest: { kernelSize: 49, sigma: 23 } }, h = { tiny: { modelFile: "tiny-640x360.onnx", inputSize: { width: 640, height: 360 }, outputSize: { width: 640, height: 360 }, upscaleSize: { width: 640, height: 360 }, generationIntervalMs: 48, modelRank: 4 }, low: { modelFile: "low-640x360.onnx", inputSize: { width: 640, height: 360 }, outputSize: { width: 640, height: 360 }, upscaleSize: { width: 640, height: 360 }, generationIntervalMs: 48, modelRank: 4 }, medium: { modelFile: "medium-640x360.onnx", inputSize: { width: 640, height: 360 }, outputSize: { width: 640, height: 360 }, upscaleSize: { width: 640, height: 360 }, generationIntervalMs: 36, modelRank: 4 }, high: { modelFile: "medium-640x360.onnx", inputSize: { width: 640, height: 360 }, outputSize: { width: 640, height: 360 }, upscaleSize: { width: 640, height: 360 }, generationIntervalMs: 24, modelRank: 4 }, ultra: { modelFile: "medium-640x360.onnx", inputSize: { width: 640, height: 360 }, outputSize: { width: 640, height: 360 }, upscaleSize: { width: 640, height: 360 }, generationIntervalMs: 0, modelRank: 4 } }, s = "${BASE_PATH_MODELS}/${MODEL_VERSION}/${MODEL_FILE}", a = "${BASE_PATH_WASM}/${ONNX_VERSION}/", d = "${BASE_PATH_WORKERS}/${LIB_VERSION}/${WORKER_FILE}", m = {
  onnxVersion: e,
  modelVersion: t,
  workerFile: i,
  baseUrls: n,
  basePaths: o,
  blurStrengths: l,
  modelQualities: h,
  modelUrlTemplate: s,
  wasmPathTemplate: a,
  workerUrlTemplate: d
};
export {
  o as basePaths,
  n as baseUrls,
  l as blurStrengths,
  m as default,
  h as modelQualities,
  s as modelUrlTemplate,
  t as modelVersion,
  e as onnxVersion,
  a as wasmPathTemplate,
  i as workerFile,
  d as workerUrlTemplate
};
