// @webex/ladon-ts/build/module/lib/config.js
import { toAbsoluteURI } from './util';
import { LIB_VERSION } from './version';
export const BASE_URL = 'https://models.intelligence.webex.com/ladon';
export const LOCAL_BASE_URL = 'https://localhost';
export const LOCAL_WORKER_URI = '/segmentation.js';
export const LOCAL_PORT = 1235;
export const MODEL_VERSION = '0.7.7.1';
export const WASM_URI = '/wasm/1.15.1b/';
export var Quality;
(function (Quality) {
    Quality["TINY"] = "TINY";
    Quality["LOW"] = "LOW";
    Quality["MEDIUM"] = "MEDIUM";
    Quality["HIGH"] = "HIGH";
    Quality["ULTRA"] = "ULTRA";
})(Quality || (Quality = {}));
export var BlurStrength;
(function (BlurStrength) {
    BlurStrength["WEAK"] = "WEAK";
    BlurStrength["MODERATE"] = "MODERATE";
    BlurStrength["STRONG"] = "STRONG";
    BlurStrength["STRONGER"] = "STRONGER";
    BlurStrength["STRONGEST"] = "STRONGEST";
})(BlurStrength || (BlurStrength = {}));
export const getBlurConfig = (blur_strength, horizontal_mirror) => {
    let kernel_size = 25;
    let sigma = 11;
    switch (blur_strength) {
        case BlurStrength.WEAK:
            kernel_size = 9;
            sigma = 2;
            break;
        case BlurStrength.MODERATE:
            kernel_size = 15;
            sigma = 6;
            break;
        case BlurStrength.STRONG:
            kernel_size = 25;
            sigma = 11;
            break;
        case BlurStrength.STRONGER:
            kernel_size = 37;
            sigma = 17;
            break;
        case BlurStrength.STRONGEST:
            kernel_size = 49;
            sigma = 23;
            break;
        default:
            break;
    }
    return {
        type: 'blur',
        horizontal_mirror,
        kernel_size,
        sigma,
    };
};
export const getPreloadConfig = (quality = Quality.LOW, generator = 'local', model_url_resolver = (model_path) => Promise.resolve(toAbsoluteURI(model_path, BASE_URL)), base_url = BASE_URL, useLocalWorker = false) => {
    const model_version = MODEL_VERSION;
    const localBaseUrl = `${LOCAL_BASE_URL}:${LOCAL_PORT}`;
    const workerBaseUrl = useLocalWorker ? localBaseUrl : base_url;
    const workerScriptPath = useLocalWorker
        ? LOCAL_WORKER_URI
        : `/${LIB_VERSION}/worker/segmentation.js`;
    const worker_uri = toAbsoluteURI(workerScriptPath, workerBaseUrl);
    const base_config = {
        base_url,
        wasm_uri: WASM_URI,
        mask: {
            generator,
            generation_interval_ms: 48,
            model_uri: `/models/${model_version}/tiny-256x144-256x144.onnx`,
            input_size: { height: 144, width: 256 },
            output_size: { height: 144, width: 256 },
            upscaled_size: { height: 360, width: 640 },
            model_rank: 4,
            model_url_resolver,
            worker_uri,
        },
    };
    let mask_quality_config = {};
    switch (quality) {
        case Quality.TINY:
            mask_quality_config = {
                generation_interval_ms: 48,
                model_uri: `/models/${model_version}/tiny-256x144-256x144.onnx`,
                input_size: { height: 144, width: 256 },
                output_size: { height: 144, width: 256 },
                upscaled_size: { height: 144, width: 256 },
                model_rank: 4,
            };
            break;
        case Quality.LOW:
            mask_quality_config = {
                generation_interval_ms: 48,
                model_uri: `/models/${model_version}/low-256x144-256x144.onnx`,
                input_size: { height: 144, width: 256 },
                output_size: { height: 144, width: 256 },
                upscaled_size: { height: 360, width: 640 },
                model_rank: 4,
            };
            break;
        case Quality.MEDIUM:
            mask_quality_config = {
                generation_interval_ms: 24,
                model_uri: `/models/${model_version}/medium-256x144-256x144.onnx`,
                input_size: { height: 144, width: 256 },
                output_size: { height: 144, width: 256 },
                upscaled_size: { height: 360, width: 640 },
                model_rank: 4,
            };
            break;
        case Quality.HIGH:
            mask_quality_config = {
                generation_interval_ms: 36,
                model_uri: `/models/${model_version}/medium-640x360-256x144.onnx`,
                input_size: { height: 360, width: 640 },
                output_size: { height: 360, width: 640 },
                upscaled_size: { height: 360, width: 640 },
                model_rank: 4,
            };
            break;
        case Quality.ULTRA:
            mask_quality_config = {
                generation_interval_ms: 0,
                model_uri: `/models/${model_version}/medium-640x360-256x144.onnx`,
                input_size: { height: 360, width: 640 },
                output_size: { height: 360, width: 640 },
                upscaled_size: { height: 360, width: 640 },
                model_rank: 4,
            };
            break;
        default:
            throw new Error('[ladon-ts] - unknown quality option: ' + quality);
    }
    return Object.assign(Object.assign({}, base_config), { mask: Object.assign(Object.assign({}, base_config.mask), mask_quality_config) });
};
export const getConfig = (height, width, quality = Quality.LOW, generator = 'local', renderConfig = {
    type: 'blur',
    horizontal_mirror: false,
    kernel_size: 37,
    sigma: 17,
}, modelUrlResolver = (model_path) => Promise.resolve(toAbsoluteURI(model_path, BASE_URL)), baseUrl = BASE_URL, useLocalWorker = false) => {
    const basePreloadConfig = getPreloadConfig(quality, generator, modelUrlResolver, baseUrl, useLocalWorker);
    const maskConfig = basePreloadConfig.mask;
    return Object.assign(Object.assign({}, basePreloadConfig), { render: renderConfig, input: { height, width }, mask: Object.assign(Object.assign({}, maskConfig), { 
            // Need to ensure we don't override the tiny model's input and output sizes.
            input_size: maskConfig.input_size || { height, width }, output_size: maskConfig.output_size || { height, width }, upscaled_size: maskConfig.upscaled_size || { height, width } }) });
};
