// @webex/calling/dist/Voicemail/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.VOICE_MESSAGING_MESSAGES = exports.VOICEMAIL_FILE = exports.VOICEMAILS = exports.VMGATEWAY = exports.TRANSCRIPT_STATUS = exports.TRANSCRIPT_CONTENT = exports.SUMMARY = exports.SORT_ORDER = exports.RADIX_RAND = exports.PREFIX = exports.OLD_URGENT_MESSAGES = exports.OLD_MESSAGES = exports.OFFSET_LIMIT = exports.OFFSET_INDEX = exports.OFFSET = exports.NO_VOICEMAIL_STATUS_CODE = exports.NO_VOICEMAIL_MSG = exports.NEW_URGENT_MESSAGES = exports.NEW_MESSAGES = exports.METHODS = exports.MESSAGE_SUMMARY = exports.MESSAGE_MEDIA_CONTENT = exports.MARK_AS_UNREAD = exports.MARK_AS_READ = exports.LIMIT = exports.JSON_FORMAT = exports.CALLS = exports.BW_TOKEN_FETCH_ENDPOINT = exports.BROADWORKS_VOICEMAIL_FILE = exports.API_V1 = void 0;
var API_V1 = exports.API_V1 = 'api/v1';
var VOICEMAIL_FILE = exports.VOICEMAIL_FILE = 'VoicemailClient';
var BROADWORKS_VOICEMAIL_FILE = exports.BROADWORKS_VOICEMAIL_FILE = 'BroadworksBackendConnector';
var CALLS = exports.CALLS = 'calls';
var BW_TOKEN_FETCH_ENDPOINT = exports.BW_TOKEN_FETCH_ENDPOINT = '/idp/bwtoken/fetch';
var JSON_FORMAT = exports.JSON_FORMAT = '?format=json';
var LIMIT = exports.LIMIT = '&limit';
var MARK_AS_READ = exports.MARK_AS_READ = 'MarkAsRead';
var MARK_AS_UNREAD = exports.MARK_AS_UNREAD = 'MarkAsUnread';
var MESSAGE_MEDIA_CONTENT = exports.MESSAGE_MEDIA_CONTENT = 'messageMediaContent';
var MESSAGE_SUMMARY = exports.MESSAGE_SUMMARY = 'MessageSummary';
var NO_VOICEMAIL_MSG = exports.NO_VOICEMAIL_MSG = 'No additional voicemails';
var NO_VOICEMAIL_STATUS_CODE = exports.NO_VOICEMAIL_STATUS_CODE = 204;
var OFFSET = exports.OFFSET = '?offset';
var OFFSET_INDEX = exports.OFFSET_INDEX = 0;
var OFFSET_LIMIT = exports.OFFSET_LIMIT = 100;
var PREFIX = exports.PREFIX = 2;
var RADIX_RAND = exports.RADIX_RAND = 36;
var SORT_ORDER = exports.SORT_ORDER = '&sortOrder';
var TRANSCRIPT_STATUS = exports.TRANSCRIPT_STATUS = 'status';
var SUMMARY = exports.SUMMARY = 'summary';
var TRANSCRIPT_CONTENT = exports.TRANSCRIPT_CONTENT = 'content';
var VMGATEWAY = exports.VMGATEWAY = 'vmgateway';
var VOICEMAILS = exports.VOICEMAILS = 'voicemails';
var VOICE_MESSAGING_MESSAGES = exports.VOICE_MESSAGING_MESSAGES = 'VoiceMessagingMessages';
var NEW_MESSAGES = exports.NEW_MESSAGES = 'newMessages';
var OLD_MESSAGES = exports.OLD_MESSAGES = 'oldMessages';
var NEW_URGENT_MESSAGES = exports.NEW_URGENT_MESSAGES = 'newUrgentMessages';
var OLD_URGENT_MESSAGES = exports.OLD_URGENT_MESSAGES = 'oldUrgentMessages';
/**
 * Method names for logging
 */
var METHODS = exports.METHODS = {
  INIT: 'init',
  GET_SDK_CONNECTOR: 'getSDKConnector',
  GET_USER_ID: 'getUserId',
  GET_BW_TOKEN: 'getBwToken',
  SET_XSI_VOICE_MESSAGE_URI: 'setXsiVoiceMessageURI',
  GET_VOICEMAIL_LIST: 'getVoicemailList',
  GET_VOICEMAIL_CONTENT: 'getVoicemailContent',
  GET_VOICEMAIL_SUMMARY: 'getVoicemailSummary',
  VOICEMAIL_MARK_AS_READ: 'voicemailMarkAsRead',
  VOICEMAIL_MARK_AS_UNREAD: 'voicemailMarkAsUnread',
  DELETE_VOICEMAIL: 'deleteVoicemail',
  GET_VM_TRANSCRIPT: 'getVMTranscript',
  RESOLVE_CONTACT: 'resolveContact',
  GET_VOICEMAIL_CONTENT_UCM: 'getVoicemailContentUcm',
  RETURN_UCM_PROMISE: 'returnUcmPromise',
  INITIALIZE_BACKEND_CONNECTOR: 'initializeBackendConnector',
  SUBMIT_METRIC: 'submitMetric',
  GET_AUTH_HEADERS: 'getAuthHeaders',
  SET_UCM_VOICE_MESSAGE_BASE_URI: 'setUcmVoiceMessageBaseURI'
};
