// @webex/calling/dist/Voicemail/WxCallBackendConnector.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.WxCallBackendConnector = void 0;
var _values = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/values"));
var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));
var _isArray = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/array/is-array"));
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _SDKConnector = _interopRequireDefault(require("../SDKConnector"));
var _constants = require("../common/constants");
var _Utils = require("../common/Utils");
var _types = require("../common/types");
var _Logger = _interopRequireDefault(require("../Logger"));
var _constants2 = require("./constants");
/* eslint-disable dot-notation */
/* eslint-disable no-underscore-dangle */
/**
 *
 */
var WxCallBackendConnector = exports.WxCallBackendConnector = /*#__PURE__*/function () {
  /**
   * @param webex - An object of the webex-js-sdk type.
   * @param logger - Logger interface.
   */
  function WxCallBackendConnector(webex, logger) {
    (0, _classCallCheck2.default)(this, WxCallBackendConnector);
    (0, _defineProperty2.default)(this, "xsiEndpoint", void 0);
    (0, _defineProperty2.default)(this, "userId", void 0);
    (0, _defineProperty2.default)(this, "context", void 0);
    (0, _defineProperty2.default)(this, "sdkConnector", void 0);
    (0, _defineProperty2.default)(this, "xsiVoiceMessageURI", void 0);
    (0, _defineProperty2.default)(this, "webex", void 0);
    this.sdkConnector = _SDKConnector.default;
    if (!this.sdkConnector.getWebex()) {
      _SDKConnector.default.setWebex(webex);
    }
    this.context = Math.random().toString(_constants2.RADIX_RAND).substring(_constants2.PREFIX);
    this.webex = this.sdkConnector.getWebex();
    this.userId = this.webex.internal.device.userId;
    _Logger.default.setLogger(logger.level, _constants.WEBEX_CALLING_CONNECTOR_FILE);
  }

  /**
   * Initializing Webex calling voicemail connector.
   *
   * @returns Response.
   */
  (0, _createClass2.default)(WxCallBackendConnector, [{
    key: "init",
    value: function init() {
      var loggerContext = {
        file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
        method: 'init'
      };
      _Logger.default.info('Initializing Webex calling voicemail connector', loggerContext);
      var response = this.setXsiVoiceMessageURI();
      return response;
    }

    /**
     * @ignore
     * SDK connector function.
     * @returns SdkConnector.
     */
  }, {
    key: "getSDKConnector",
    value: function getSDKConnector() {
      return this.sdkConnector;
    }

    /**
     * Register XSI URL.
     */
  }, {
    key: "setXsiVoiceMessageURI",
    value: (function () {
      var _setXsiVoiceMessageURI = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var responseDetails, loggerContext;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'setXsiVoiceMessageURI'
              };
              _context.next = 3;
              return (0, _Utils.getXsiActionEndpoint)(this.webex, loggerContext, _types.CALLING_BACKEND.WXC);
            case 3:
              this.xsiEndpoint = _context.sent;
              _Logger.default.info("XsiEndpoint is ".concat(this.xsiEndpoint), loggerContext);
              if (this.userId) {
                this.xsiVoiceMessageURI = "".concat(this.xsiEndpoint, "/").concat(_constants.BW_XSI_ENDPOINT_VERSION, "/").concat(_constants.USER, "/").concat(this.userId, "/").concat(_constants2.VOICE_MESSAGING_MESSAGES);
                responseDetails = {
                  statusCode: _constants.SUCCESS_STATUS_CODE,
                  data: {},
                  message: _constants.SUCCESS_MESSAGE
                };
              }
              return _context.abrupt("return", responseDetails);
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function setXsiVoiceMessageURI() {
        return _setXsiVoiceMessageURI.apply(this, arguments);
      }
      return setXsiVoiceMessageURI;
    }()
    /**
     * Fetch voicemail list for Webex users.
     *
     * @param sort - Sort voicemail list (ASC | DESC).
     * @param offset - Number of records to skip.
     * @param offsetLimit - Number of voicemail list to fetch from the offset.
     * @param refresh - Refresh the list of voicemails from backend.
     * @returns Promise.
     */
    )
  }, {
    key: "getVoicemailList",
    value: (function () {
      var _getVoicemailList = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(offset, offsetLimit, sort, refresh) {
        var loggerContext, messageinfo, urlXsi, sortParam, _voicemailListRespons, _voicemailListRespons2, _voicemailListRespons3, response, voicemailListResponse, _voicemailListRespons4, _voicemailListRespons5, _voicemailListRespons6, _voicemailListRespons7, errorInfo, errorStatus, _fetchVoicemailList, messages, moreVMAvailable, responseDetails;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'getVoicemailList'
              };
              _Logger.default.info("Offset: ".concat(offset, " Offset limit: ").concat(offsetLimit, " Sort type:").concat(sort), loggerContext);
              if (!refresh) {
                _context2.next = 19;
                break;
              }
              urlXsi = "".concat(this.xsiVoiceMessageURI).concat(_constants2.JSON_FORMAT);
              sortParam = (0, _values.default)(_types.SORT).includes(sort) ? sort : _types.SORT.DEFAULT;
              _context2.prev = 5;
              _context2.next = 8;
              return this.webex.request({
                uri: "".concat(urlXsi),
                method: _types.HTTP_METHODS.GET
              });
            case 8:
              response = _context2.sent;
              voicemailListResponse = response.body;
              if ((0, _keys.default)(voicemailListResponse === null || voicemailListResponse === void 0 ? void 0 : (_voicemailListRespons = voicemailListResponse.VoiceMessagingMessages) === null || _voicemailListRespons === void 0 ? void 0 : _voicemailListRespons.messageInfoList).length === 0) {
                messageinfo = [];
              } else if (!(0, _isArray.default)(voicemailListResponse === null || voicemailListResponse === void 0 ? void 0 : (_voicemailListRespons2 = voicemailListResponse.VoiceMessagingMessages) === null || _voicemailListRespons2 === void 0 ? void 0 : (_voicemailListRespons3 = _voicemailListRespons2.messageInfoList) === null || _voicemailListRespons3 === void 0 ? void 0 : _voicemailListRespons3.messageInfo)) {
                messageinfo = Array(voicemailListResponse === null || voicemailListResponse === void 0 ? void 0 : (_voicemailListRespons4 = voicemailListResponse.VoiceMessagingMessages) === null || _voicemailListRespons4 === void 0 ? void 0 : (_voicemailListRespons5 = _voicemailListRespons4.messageInfoList) === null || _voicemailListRespons5 === void 0 ? void 0 : _voicemailListRespons5.messageInfo);
              } else {
                messageinfo = voicemailListResponse === null || voicemailListResponse === void 0 ? void 0 : (_voicemailListRespons6 = voicemailListResponse.VoiceMessagingMessages) === null || _voicemailListRespons6 === void 0 ? void 0 : (_voicemailListRespons7 = _voicemailListRespons6.messageInfoList) === null || _voicemailListRespons7 === void 0 ? void 0 : _voicemailListRespons7.messageInfo;
                messageinfo = (0, _Utils.getSortedVoicemailList)(messageinfo, sortParam);
              }
              (0, _Utils.storeVoicemailList)(this.context, messageinfo);
              _context2.next = 19;
              break;
            case 14:
              _context2.prev = 14;
              _context2.t0 = _context2["catch"](5);
              errorInfo = _context2.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context2.abrupt("return", errorStatus);
            case 19:
              _fetchVoicemailList = (0, _Utils.fetchVoicemailList)(this.context, offset, offsetLimit, loggerContext), messages = _fetchVoicemailList.messages, moreVMAvailable = _fetchVoicemailList.moreVMAvailable;
              responseDetails = {
                statusCode: moreVMAvailable ? _constants.SUCCESS_STATUS_CODE : _constants2.NO_VOICEMAIL_STATUS_CODE,
                data: {
                  voicemailList: messages
                },
                message: moreVMAvailable ? _constants.SUCCESS_MESSAGE : _constants2.NO_VOICEMAIL_MSG
              };
              return _context2.abrupt("return", responseDetails);
            case 22:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this, [[5, 14]]);
      }));
      function getVoicemailList(_x, _x2, _x3, _x4) {
        return _getVoicemailList.apply(this, arguments);
      }
      return getVoicemailList;
    }()
    /**
     * Fetch the voicemail contents for the messageId.
     *
     * @param messageId -string result from the voicemail list.
     * @returns Promise.
     */
    )
  }, {
    key: "getVoicemailContent",
    value: (function () {
      var _getVoicemailContent = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3(messageId) {
        var loggerContext, _mediaDetails$childNo, _mediaDetails$childNo2, voicemailContentUrl, response, parser, xmlDOM, mediaDetails, mediaType, mediaContent, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'getVoicemailContent'
              };
              _context3.prev = 1;
              voicemailContentUrl = "".concat(this.xsiEndpoint).concat(messageId);
              _context3.next = 5;
              return this.webex.request({
                uri: "".concat(voicemailContentUrl),
                method: _types.HTTP_METHODS.GET
              });
            case 5:
              response = _context3.sent;
              parser = new DOMParser();
              xmlDOM = parser.parseFromString(response[_constants.RAW_REQUEST].response, _constants.XML_TYPE);
              mediaDetails = xmlDOM.getElementsByTagName(_constants2.MESSAGE_MEDIA_CONTENT)[0];
              mediaType = (_mediaDetails$childNo = mediaDetails.childNodes[1]) === null || _mediaDetails$childNo === void 0 ? void 0 : _mediaDetails$childNo.textContent;
              mediaContent = (_mediaDetails$childNo2 = mediaDetails.childNodes[2]) === null || _mediaDetails$childNo2 === void 0 ? void 0 : _mediaDetails$childNo2.textContent;
              _Logger.default.info("Media type is  ".concat(mediaType), loggerContext);
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {
                  voicemailContent: {
                    type: mediaType,
                    content: mediaContent
                  }
                },
                message: _constants.SUCCESS_MESSAGE
              };
              return _context3.abrupt("return", responseDetails);
            case 16:
              _context3.prev = 16;
              _context3.t0 = _context3["catch"](1);
              errorInfo = _context3.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              _Logger.default.info("Voice mail content error is ".concat(errorStatus), loggerContext);
              return _context3.abrupt("return", errorStatus);
            case 22:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this, [[1, 16]]);
      }));
      function getVoicemailContent(_x5) {
        return _getVoicemailContent.apply(this, arguments);
      }
      return getVoicemailContent;
    }()
    /**
     * Fetches a quantitative summary of voicemails for a user.
     *
     * @returns - A Promise that resolves with the data containing counters for newMessages, oldMessage, newUrgentMessages, oldUrgentMessages.
     */
    )
  }, {
    key: "getVoicemailSummary",
    value: (function () {
      var _getVoicemailSummary = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4() {
        var loggerContext, voicemailSummaryUrl, response, parser, xmlDOM, voicemailSummary, newMessages, newUrgentMessages, oldMessages, oldUrgentMessages, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'getVoicemailSummary'
              };
              _context4.prev = 1;
              voicemailSummaryUrl = "".concat(this.xsiEndpoint, "/").concat(_constants.BW_XSI_ENDPOINT_VERSION, "/").concat(_constants.USER, "/").concat(this.userId, "/").concat(_constants2.CALLS, "/").concat(_constants2.MESSAGE_SUMMARY);
              _context4.next = 5;
              return this.webex.request({
                uri: "".concat(voicemailSummaryUrl),
                method: _types.HTTP_METHODS.GET
              });
            case 5:
              response = _context4.sent;
              parser = new DOMParser();
              xmlDOM = parser.parseFromString(response[_constants.RAW_REQUEST].response, _constants.XML_TYPE);
              voicemailSummary = xmlDOM.getElementsByTagName(_constants2.SUMMARY)[0];
              newMessages = voicemailSummary.getElementsByTagName(_constants2.NEW_MESSAGES)[0];
              newUrgentMessages = voicemailSummary.getElementsByTagName(_constants2.NEW_URGENT_MESSAGES)[0];
              oldMessages = voicemailSummary.getElementsByTagName(_constants2.OLD_MESSAGES)[0];
              oldUrgentMessages = voicemailSummary.getElementsByTagName(_constants2.OLD_URGENT_MESSAGES)[0];
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {
                  voicemailSummary: {
                    newMessages: newMessages ? Number(newMessages.textContent) : 0,
                    newUrgentMessages: newUrgentMessages ? Number(newUrgentMessages.textContent) : 0,
                    oldMessages: oldMessages ? Number(oldMessages.textContent) : 0,
                    oldUrgentMessages: oldUrgentMessages ? Number(oldUrgentMessages.textContent) : 0
                  }
                },
                message: _constants.SUCCESS_MESSAGE
              };
              return _context4.abrupt("return", responseDetails);
            case 17:
              _context4.prev = 17;
              _context4.t0 = _context4["catch"](1);
              errorInfo = _context4.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              _Logger.default.error(new Error("Voicemail summary error is ".concat(errorStatus)), loggerContext);
              return _context4.abrupt("return", errorStatus);
            case 23:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this, [[1, 17]]);
      }));
      function getVoicemailSummary() {
        return _getVoicemailSummary.apply(this, arguments);
      }
      return getVoicemailSummary;
    }()
    /**
     * Fetch voicemail read message status for the messageId.
     *
     * @param messageId -string result from the voicemail list.
     * @returns Promise.
     */
    )
  }, {
    key: "voicemailMarkAsRead",
    value: (function () {
      var _voicemailMarkAsRead = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee5(messageId) {
        var loggerContext, voicemailContentUrl, response, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'voicemailMarkAsRead'
              };
              _context5.prev = 1;
              voicemailContentUrl = "".concat(this.xsiEndpoint).concat(messageId, "/").concat(_constants2.MARK_AS_READ);
              _context5.next = 5;
              return this.webex.request({
                uri: voicemailContentUrl,
                method: _types.HTTP_METHODS.PUT
              });
            case 5:
              response = _context5.sent;
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {},
                message: _constants.SUCCESS_MESSAGE
              };
              return _context5.abrupt("return", responseDetails);
            case 10:
              _context5.prev = 10;
              _context5.t0 = _context5["catch"](1);
              errorInfo = _context5.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context5.abrupt("return", errorStatus);
            case 15:
            case "end":
              return _context5.stop();
          }
        }, _callee5, this, [[1, 10]]);
      }));
      function voicemailMarkAsRead(_x6) {
        return _voicemailMarkAsRead.apply(this, arguments);
      }
      return voicemailMarkAsRead;
    }()
    /**
     * Fetch voicemail unread message status for the messageId.
     *
     * @param messageId -string result from the voicemail list.
     * @returns Promise.
     */
    )
  }, {
    key: "voicemailMarkAsUnread",
    value: (function () {
      var _voicemailMarkAsUnread = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee6(messageId) {
        var loggerContext, voicemailContentUrl, response, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'voicemailMarkAsUnread'
              };
              _context6.prev = 1;
              voicemailContentUrl = "".concat(this.xsiEndpoint).concat(messageId, "/").concat(_constants2.MARK_AS_UNREAD);
              _context6.next = 5;
              return this.webex.request({
                uri: voicemailContentUrl,
                method: _types.HTTP_METHODS.PUT
              });
            case 5:
              response = _context6.sent;
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {},
                message: _constants.SUCCESS_MESSAGE
              };
              return _context6.abrupt("return", responseDetails);
            case 10:
              _context6.prev = 10;
              _context6.t0 = _context6["catch"](1);
              errorInfo = _context6.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context6.abrupt("return", errorStatus);
            case 15:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this, [[1, 10]]);
      }));
      function voicemailMarkAsUnread(_x7) {
        return _voicemailMarkAsUnread.apply(this, arguments);
      }
      return voicemailMarkAsUnread;
    }()
    /**
     * Fetch voicemail delete status for the messageId.
     *
     * @param messageId -string result from the voicemail list.
     * @returns Promise.
     */
    )
  }, {
    key: "deleteVoicemail",
    value: (function () {
      var _deleteVoicemail = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee7(messageId) {
        var loggerContext, voicemailContentUrl, response, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee7$(_context7) {
          while (1) switch (_context7.prev = _context7.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'deleteVoicemail'
              };
              _context7.prev = 1;
              voicemailContentUrl = "".concat(this.xsiEndpoint).concat(messageId);
              _context7.next = 5;
              return this.webex.request({
                uri: voicemailContentUrl,
                method: _types.HTTP_METHODS.DELETE
              });
            case 5:
              response = _context7.sent;
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {},
                message: _constants.SUCCESS_MESSAGE
              };
              return _context7.abrupt("return", responseDetails);
            case 10:
              _context7.prev = 10;
              _context7.t0 = _context7["catch"](1);
              errorInfo = _context7.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context7.abrupt("return", errorStatus);
            case 15:
            case "end":
              return _context7.stop();
          }
        }, _callee7, this, [[1, 10]]);
      }));
      function deleteVoicemail(_x8) {
        return _deleteVoicemail.apply(this, arguments);
      }
      return deleteVoicemail;
    }()
    /**
     * Fetch voicemail transcript for the messageId.
     *
     * @param messageId - MessageId to fetch voicemail transcript.
     * @returns Promise.
     */
    )
  }, {
    key: "getVMTranscript",
    value: (function () {
      var _getVMTranscript = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee8(messageId) {
        var loggerContext, voicemailContentUrl, response, parser, xmlDOM, status, transcript, responseDetails, errorInfo, errorStatus;
        return _regenerator.default.wrap(function _callee8$(_context8) {
          while (1) switch (_context8.prev = _context8.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: 'getVMTranscript'
              };
              _context8.prev = 1;
              voicemailContentUrl = "".concat(this.xsiEndpoint).concat(messageId, "/").concat(_constants.TRANSCRIPT);
              _context8.next = 5;
              return this.webex.request({
                uri: voicemailContentUrl,
                method: _types.HTTP_METHODS.GET
              });
            case 5:
              response = _context8.sent;
              parser = new DOMParser();
              xmlDOM = parser.parseFromString(response[_constants.RAW_REQUEST].response, _constants.XML_TYPE);
              status = xmlDOM.getElementsByTagName(_constants2.TRANSCRIPT_STATUS)[0];
              transcript = xmlDOM.getElementsByTagName(_constants2.TRANSCRIPT_CONTENT)[0];
              responseDetails = {
                statusCode: Number(response.statusCode),
                data: {
                  voicemailTranscript: transcript === null || transcript === void 0 ? void 0 : transcript.textContent
                },
                message: status.textContent
              };
              return _context8.abrupt("return", responseDetails);
            case 14:
              _context8.prev = 14;
              _context8.t0 = _context8["catch"](1);
              errorInfo = _context8.t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context8.abrupt("return", errorStatus);
            case 19:
            case "end":
              return _context8.stop();
          }
        }, _callee8, this, [[1, 14]]);
      }));
      function getVMTranscript(_x9) {
        return _getVMTranscript.apply(this, arguments);
      }
      return getVMTranscript;
    }()
    /**
     * Resolve the Contact from userId or display name.
     *
     * @param callingPartyInfo - Calling Party Info.
     */
    )
  }, {
    key: "resolveContact",
    value: function resolveContact(callingPartyInfo) {
      return (0, _Utils.resolveContact)(callingPartyInfo);
    }
  }]);
  return WxCallBackendConnector;
}();
