// @webex/ladon-ts/build/module/lib/renderer/blur-renderer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { bilinearUpsample, compileShader, linkProgram } from '../util';
import { WebGLRenderer } from './webgl-renderer';
export class BlurRenderer extends WebGLRenderer {
    constructor(output_canvas, config) {
        super(output_canvas, config);
        this.output_canvas = output_canvas;
        this.config = config;
        this.fragment_shader_common = `#version 300 es
precision highp float;

#define KERNEL_SIZE     ${this.config.render.kernel_size}
#define HORIZONTAL      vec2(1.0, 0.0)
#define VERTICAL        vec2(0.0, 1.0)

out vec4 outColor;

uniform sampler2D       u_image;
uniform sampler2D       u_mask;
uniform vec2            u_resolution;
uniform float           u_kernel[KERNEL_SIZE];

// Function to retrieve the mask value
float get_mask(vec2 st) {
    return texture(u_mask, st).r;
}

// Get the inverse of the masked area
float get_inv_mask(vec2 st) {
    return 1.0 - get_mask(st);
}

// Blurring function
vec3 blur(sampler2D src, vec2 uv, vec2 direction) {
    vec4 accum = vec4(0.0);
    vec2 step = direction / u_resolution;
    int offset = (KERNEL_SIZE - 1) / 2;

    float weightSum = 0.0;
    for (int i = 0; i < KERNEL_SIZE; i++) {
        vec2 sampleUV = uv + step * float(i - offset);
        vec4 sampleTex = texture(src, sampleUV);
        float weight = u_kernel[i];
        accum += sampleTex * weight;
        weightSum += weight;
    }

    weightSum = max(weightSum, 0.001); // Avoid division by zero
    return accum.rgb / weightSum;
}
`;
        this.fragment_shader_pass_1_src = this.fragment_shader_common +
            `
void main() {
    vec2 st = gl_FragCoord.xy / u_resolution;
    outColor = vec4(blur(u_image, st, HORIZONTAL), 1.0);
}
`;
        this.fragment_shader_pass_2_src = this.fragment_shader_common +
            `
uniform sampler2D   u_blur;

void main() {
    ${this.getUVCalculation()};

    vec3 blurred = blur(u_blur, st, VERTICAL);
    vec3 original = texture(u_image, st).rgb;
    float mask = get_inv_mask(st); // Use inverse mask to blur background, not user

    // Correct blending based on the mask
    vec3 color = mix(original, blurred, mask); // Ensure mask properly determines blur influence on background
    outColor = vec4(color, 1.0);
}
`;
        if (this.config.render.kernel_size % 2 === 0) {
            throw new Error('[ladon-ts] - kernel_size must be odd');
        }
        // program pass 1 setup
        {
            const vertex_shader = compileShader(this.gl, this.gl.VERTEX_SHADER, this.vertex_shader_src);
            const fragment_shader = compileShader(this.gl, this.gl.FRAGMENT_SHADER, this.fragment_shader_pass_1_src);
            this.program_1 = linkProgram(this.gl, vertex_shader, fragment_shader);
            this.gl.useProgram(this.program_1);
            const framebuffer = this.gl.createFramebuffer();
            if (!framebuffer) {
                throw new Error('[ladon-ts] - error creating framebuffer');
            }
            this.framebuffer = framebuffer;
            const transfer_texture = this.gl.createTexture();
            if (!transfer_texture) {
                throw new Error('[ladon-ts] - error creating output texture');
            }
            this.transfer_texture = transfer_texture;
            this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, this.framebuffer);
            this.gl.bindTexture(this.gl.TEXTURE_2D, this.transfer_texture);
            this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_MIN_FILTER, this.gl.LINEAR);
            this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_S, this.gl.CLAMP_TO_EDGE);
            this.gl.texParameteri(this.gl.TEXTURE_2D, this.gl.TEXTURE_WRAP_T, this.gl.CLAMP_TO_EDGE);
            this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.RGBA, this.config.input.width, this.config.input.height, 0, this.gl.RGBA, this.gl.UNSIGNED_BYTE, null);
            this.gl.framebufferTexture2D(this.gl.FRAMEBUFFER, this.gl.COLOR_ATTACHMENT0, this.gl.TEXTURE_2D, this.transfer_texture, 0);
            this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, null);
            this.setSharedUniforms(this.program_1);
            const u_kernel = this.gl.getUniformLocation(this.program_1, 'u_kernel');
            if (!u_kernel) {
                throw new Error('[ladon-ts] - error finding "u_kernel" uniform');
            }
            this.gl.uniform1fv(u_kernel, this.getBlurKernel());
        }
        // program pass 2 setup
        {
            const vertex_shader = compileShader(this.gl, this.gl.VERTEX_SHADER, this.vertex_shader_src);
            const fragment_shader = compileShader(this.gl, this.gl.FRAGMENT_SHADER, this.fragment_shader_pass_2_src);
            this.program_2 = linkProgram(this.gl, vertex_shader, fragment_shader);
            this.gl.useProgram(this.program_2);
            const u_blur = this.gl.getUniformLocation(this.program_2, 'u_blur');
            if (!u_blur) {
                throw new Error('[ladon-ts] - error finding "u_blur" uniform');
            }
            this.u_blur = u_blur;
            this.gl.activeTexture(this.gl.TEXTURE2);
            this.gl.bindTexture(this.gl.TEXTURE_2D, this.transfer_texture);
            this.gl.uniform1i(this.u_blur, 2);
            this.setSharedUniforms(this.program_2);
            const u_kernel = this.gl.getUniformLocation(this.program_2, 'u_kernel');
            if (!u_kernel) {
                throw new Error('[ladon-ts] - error finding "u_kernel" uniform');
            }
            this.gl.uniform1fv(u_kernel, this.getBlurKernel());
        }
    }
    render(input, mask) {
        return __awaiter(this, void 0, void 0, function* () {
            this.uploadVideoStream(input);
            // The mask may not be included if it has not been generated (e.g., frame skipped).
            if (mask === null || mask === void 0 ? void 0 : mask.warm) {
                const maskData = bilinearUpsample(mask.data, this.config.mask.output_size.width, this.config.mask.output_size.height, this.config.mask.upscaled_size.width, this.config.mask.upscaled_size.height);
                this.gl.activeTexture(this.gl.TEXTURE1);
                this.gl.bindTexture(this.gl.TEXTURE_2D, this.mask_texture);
                this.gl.texImage2D(this.gl.TEXTURE_2D, 0, this.gl.R8, this.config.mask.upscaled_size.width, this.config.mask.upscaled_size.height, 0, this.gl.RED, this.gl.UNSIGNED_BYTE, maskData);
            }
            // program pass 1 - blur in one direction & output to framebuffer
            this.gl.useProgram(this.program_1);
            this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, this.framebuffer);
            this.gl.drawArrays(this.gl.TRIANGLES, 0, this.verts.length / 3);
            this.gl.bindFramebuffer(this.gl.FRAMEBUFFER, null);
            // program pass 2 - blur in other direction & draw to screen
            this.gl.useProgram(this.program_2);
            this.gl.drawArrays(this.gl.TRIANGLES, 0, this.verts.length / 3);
            this.postDraw();
        });
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            // empty
        });
    }
    destroy() {
        // empty
    }
    getBlurKernel() {
        const kernel = [];
        let sum = 0;
        for (let i = 0; i < this.config.render.kernel_size; i++) {
            let w = i - (this.config.render.kernel_size - 1) / 2;
            w = Math.exp(-(Math.pow(w, 2)) / Math.pow(this.config.render.sigma, 2));
            kernel.push(w);
            sum += w;
        }
        for (let i = 0; i < this.config.render.kernel_size; i++) {
            kernel[i] = kernel[i] / sum;
        }
        return new Float32Array(kernel);
    }
}
