// @webex/calling/dist/Contacts/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.encryptedFields = exports.USERS = exports.SCIM_ID_FILTER = exports.OR = exports.METHODS = exports.GROUP_FILTER = exports.ENCRYPT_FILTER = exports.DEFAULT_GROUP_NAME = exports.CONTACT_FILTER = exports.CONTACTS_SCHEMA = exports.CONTACTS_FILE = void 0;
var CONTACTS_FILE = exports.CONTACTS_FILE = 'Contacts';
var CONTACT_FILTER = exports.CONTACT_FILTER = 'contacts';
var GROUP_FILTER = exports.GROUP_FILTER = 'groups';
var ENCRYPT_FILTER = exports.ENCRYPT_FILTER = 'encrypt';
var USERS = exports.USERS = 'Users';
var DEFAULT_GROUP_NAME = exports.DEFAULT_GROUP_NAME = 'Other contacts';
var CONTACTS_SCHEMA = exports.CONTACTS_SCHEMA = 'urn:cisco:codev:identity:contact:core:1.0';
var SCIM_ID_FILTER = exports.SCIM_ID_FILTER = 'id eq';
var OR = exports.OR = ' or ';
var encryptedFields = exports.encryptedFields = /*#__PURE__*/function (encryptedFields) {
  encryptedFields["ADDRESS_INFO"] = "addressInfo";
  encryptedFields["AVATAR_URL"] = "avatarURL";
  encryptedFields["COMPANY"] = "companyName";
  encryptedFields["DISPLAY_NAME"] = "displayName";
  encryptedFields["EMAILS"] = "emails";
  encryptedFields["FIRST_NAME"] = "firstName";
  encryptedFields["LAST_NAME"] = "lastName";
  encryptedFields["PHONE_NUMBERS"] = "phoneNumbers";
  encryptedFields["SIP_ADDRESSES"] = "sipAddresses";
  encryptedFields["TITLE"] = "title";
  return encryptedFields;
}({});
var METHODS = exports.METHODS = {
  GET_CONTACTS: 'getContacts',
  CREATE_NEW_ENCRYPTION_KEY_URL: 'createNewEncryptionKeyUrl',
  FETCH_ENCRYPTION_KEY_URL: 'fetchEncryptionKeyUrl',
  FETCH_DEFAULT_GROUP: 'fetchDefaultGroup',
  CREATE_CONTACT_GROUP: 'createContactGroup',
  DELETE_CONTACT_GROUP: 'deleteContactGroup',
  CREATE_CONTACT: 'createContact',
  DELETE_CONTACT: 'deleteContact'
};
