// @webex/calling/dist/Contacts/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupType = exports.ContactType = void 0;
var ContactType = exports.ContactType = /*#__PURE__*/function (ContactType) {
  ContactType["CUSTOM"] = "CUSTOM";
  ContactType["CLOUD"] = "CLOUD";
  return ContactType;
}({});
/**
 * `Contact` object is used to represent a contact.
 */
var GroupType = exports.GroupType = /*#__PURE__*/function (GroupType) {
  GroupType["NORMAL"] = "NORMAL";
  GroupType["EXTERNAL"] = "EXTERNAL";
  return GroupType;
}({});
/**
 * `ContactGroup` object is used to represent a contact group.
 */
/**
 * Interface for Contacts Module.
 * This encompasses a set of APIs that enable the fetching, creation, and updating of contacts and groups.
 * These APIs return promises that resolve to a `ContactResponse` object, which contains a status code, data, and message.
 * The data field within this response object holds the contacts and groups object.
 *
 * Example - ContactResponse
 *
 * ```json
 * {
 *  statusCode: 200,
 *  data: {
 *    contacts: Contact[],
 *    groups: ContactGroup[],
 *  },
 *  message: null
 * }
 * ````
 */
